---
title: "marie_code"
output: html_document
---
--

title: "plot"

author: "Marie Peras + Shoko Iwai"

date: "10/3/2020"

output: html_document

---


This Rmd is used to generate all the figures for the multi-omics manuscript. All the data are located in /Volumes/GoogleDrive/Shared drives/M3-shared/publications/Omics_integration/data_for_the_ms/


This rmd is generating figures based on previous results: significant features (after QC) that are relevant for the manuscript are plotted here. 



```{r}

knitr::opts_chunk$set(echo=TRUE, warning=TRUE, message=FALSE, cache=FALSE, fig.path='./Figure_pngs/') 


library(ggplot2)

#library(secondgenomeR)

library(dplyr)

library(DT)

library(tibble)

library(data.table)

library(tidygraph)

library(ggraph)

library(patchwork)

library(cowplot)

library(magrittr)

library(ggrepel)

```


# Figure 1: Comparison of the gut microbiome of subjects with ASD and their NT siblings using multi-omics analysis


## Figure 1.A: Abundance of strain t__28462 in ASD vs NT subjects in 16S PC, 16S V4 and MTG Sourdough data sets


```{r}

### 16S V4

normalize <- function(ps, method, log = FALSE){
  if(method == "tss"){
    norm_tab <- apply(ps@otu_table, 2, function(x) return(x / sum(x)))
    norm_tab <- data.frame(norm_tab)
    rownames(norm_tab) <- taxa_names(ps)
    colnames(norm_tab) <- sample_names(ps)
    
  }
  if(log){
    norm_tab <- log(norm_tab)
  }
  ps@otu_table <- otu_table(norm_tab, taxa_are_rows = T)
  return(ps)
}

ps_not_norm_age_filter_complete_family <- readRDS("G:/Shared drives/M3-shared/Data/V4/200312_ASVdata8_updateAsteria/Phyloseq_objects/age_filtered_PS_objects/ps_not_norm_age_filter_complete_family.rds")


ps_not_norm_age_filter_complete_family <- normalize(ps_not_norm_age_filter_complete_family , method = "tss", log = FALSE)

ASV1597 <- 't__28462'


keep <- c(ASV1597)

ps_not_norm_age_filter_complete_family  <- prune_taxa(taxa_names(ps_not_norm_age_filter_complete_family) %in% keep, ps_not_norm_age_filter_complete_family)


plot <- psmelt(ps_not_norm_age_filter_complete_family)

plot <- plot %>% subset(select=c(OTU, Host.disease.status, Within.study.sampling.date, Abundance))
plot$data_set <- rep('16S V4', nrow(plot))


plot <- as.data.table(plot)

          # add very small value, min/100000 to 0

          plot$Abundance <- plot$Abundance+min(plot[Abundance!=0]$Abundance)/100000


# relabel

plot[, Host.disease.status:=ifelse(Host.disease.status=="Case (disease status)", "ASD", "NT")]

          

p1 <- ggplot(data=plot, aes_string(x='Host.disease.status', y = 'Abundance', fill='Host.disease.status')) +

            geom_boxplot(outlier.color=NA) + 

            geom_jitter(position=position_jitterdodge(0.2), color="gray44") + scale_y_log10() + 

            scale_fill_manual(values=c("#84CF04","#01B5BB"))+

            facet_wrap(~Within.study.sampling.date)+

            theme_bw() +

            theme(text = element_text(size=10), legend.position="none") +ylab("Relative abundance of\nt__28462") + xlab('')+ggtitle("16S-V4")

p1
```



```{r}

#### 16S PC

PS_16S_hybscore_age_filtered <- readRDS("G://Shared drives/M3-shared/Data/16S_PC/Step3_Making_phyloseq/strain_probes_approach/age_filtered_PS_objects/PS_16S_hybscore_age_filtered.rds")


PS_16S_hybscore_age_filtered  <- prune_taxa(taxa_names(PS_16S_hybscore_age_filtered) %in% c('171'), PS_16S_hybscore_age_filtered)


plot2 <- psmelt(PS_16S_hybscore_age_filtered)


plot2 <- plot2 %>% filter(Within.study.sampling.date == 'Timepoint 1') %>% mutate(OTU = case_when(OTU == '171' ~ 't__28462')) %>% subset(select=c(OTU, Host.disease.status, Within.study.sampling.date, Abundance)) %>% add_column(data_set = '16S PC')


plot2 <- as.data.table(plot2)

          # add very small value, min/100000 to 0

          plot2$Abundance <- plot2$Abundance+min(plot2[Abundance!=0]$Abundance)/100000

# relabel

plot2[, Host.disease.status:=ifelse(Host.disease.status=="Case (disease status)", "ASD", "NT")]

          

p2 <- ggplot(data=plot2, aes_string(x='Host.disease.status', y = 'Abundance', fill='Host.disease.status')) +

            geom_boxplot(outlier.color=NA) +

            geom_jitter(position=position_jitterdodge(0.2), color="gray44") + scale_y_log10() + 

            scale_fill_manual(values=c("#84CF04","#01B5BB"))+

            theme_bw() +

            theme(text = element_text(size=10), legend.position="none") +ylab("Fluorescence intensity of\nt__28462 (arbitrary unit)") + xlab('')+ggtitle("16S-PC")

p2
```



```{r}

#### MTG Sourdough

ps_sourdough_MTG_age_filtered_paired_family <- readRDS("/Volumes/GoogleDrive/Shared drives/M3-shared/Data/MTG/Step3_Making_phyloseq/age_filtered_PS_objects/ps_sourdough_MTG_age_filtered_paired_family.rds")


ps_sourdough_MTG_age_filtered_paired_family  <- prune_taxa(taxa_names(ps_sourdough_MTG_age_filtered_paired_family) %in% c('t__28462'), ps_sourdough_MTG_age_filtered_paired_family)


plot3 <- psmelt(ps_sourdough_MTG_age_filtered_paired_family)


plot3 <- plot3 %>% filter(Within.study.sampling.date == 'Timepoint 1')  %>% subset(select=c(OTU, Host.disease.status, Within.study.sampling.date, Abundance)) %>% add_column(data_set = 'MTG Sourdough')


plot3 <- as.data.table(plot3)

          # add very small value, min/100000 to 0

          plot3$Abundance <- plot3$Abundance+min(plot3[Abundance!=0]$Abundance)/100000


# relabel

plot3[, Host.disease.status:=ifelse(Host.disease.status=="Case (disease status)", "ASD", "NT")]

          

p3 <- ggplot(data=plot3, aes_string(x='Host.disease.status', y = 'Abundance', fill='Host.disease.status')) +

            geom_boxplot(outlier.color=NA) + 

            geom_jitter(position=position_jitterdodge(0.2), color="gray44") + scale_y_log10() + 

            scale_fill_manual(values=c("#84CF04","#01B5BB"))+

            theme_bw() +

            theme(text = element_text(size=10), legend.position="none") +ylab("Relative abundance of\nt__28462 ") + xlab('')+ggtitle("MTG-Sour")

```


```{r fig.width=12, fig.height=4}

Fig1A = p1 + p2 + p3 + plot_layout(widths = c(1.4, 1, 1))

```



## Figure 1.B: Abundance of Blautia A wexelerea in ASD vs NT subjects in 16S V4 and 16S PC


```{r}

# 16S V4

Compo_species <- read.csv("/Volumes/GoogleDrive/Shared drives/M3-shared/publications/Omics_integration/data_for_the_ms/16S V4/All time points/HDS/Complete_family/16S V4_Composition_Species2021-01-08.csv")

  

compo_blautia <- Compo_species %>% filter(Species == 'Blautia_A__wexlerae') 


compo_blautia <- as.data.table(compo_blautia)

          compo_blautia$Abundance <- compo_blautia$Abundance+min(compo_blautia[Abundance!=0]$Abundance)/100000

          

# relabel

compo_blautia[, Host.disease.status:=ifelse(Host.disease.status=="Case (disease status)", "ASD", "NT")]

          

p4 <- ggplot(compo_blautia, aes_string(x='Host.disease.status', y = 'Abundance' , fill='Host.disease.status')) +

            geom_boxplot(outlier.color=NA) + 

            geom_jitter(position=position_jitterdodge(0.2), color="gray44") + scale_y_log10() + 

            scale_fill_manual(values=c("#84CF04","#01B5BB"))+

            facet_wrap(~Within.study.sampling.date)+

            theme_bw() +

            theme(text = element_text(size=10), legend.position="none") +ylab("Relative abundance of\nBlautia A wexlerae") + xlab('')+ggtitle("16S-V4")

```


```{r}

# Blautia Wexlerea in 16S PC

PC_StrainApproach_HDS_Composition_Species2020_07_22 <- read.csv("/Volumes/GoogleDrive/Shared drives/M3-shared/publications/Omics_integration/data_for_the_ms/16S PC/Strain approach/Host.disease.status/16SPC_StrainApproach_HDS__Composition_Species2020-07-22.csv")


compo_blautia <- PC_StrainApproach_HDS_Composition_Species2020_07_22 %>% filter(Species == 'Blautia_A__wexlerae') 


compo_blautia <- as.data.table(compo_blautia)

          compo_blautia$Abundance <- compo_blautia$Abundance+min(compo_blautia[Abundance!=0]$Abundance)/100000

          

# relabel

compo_blautia[, Host.disease.status:=ifelse(Host.disease.status=="Case (disease status)", "ASD", "NT")]

          

p5 <- ggplot(compo_blautia, aes_string(x='Host.disease.status', y = 'Abundance' , fill='Host.disease.status')) +

            geom_boxplot(outlier.color=NA) + 

            geom_jitter(position=position_jitterdodge(0.2), color="gray44") + scale_y_log10() + 

            scale_fill_manual(values=c("#84CF04","#01B5BB"))+

            theme_bw() +

            theme(text = element_text(size=10), legend.position="none") +ylab("Fluorescence intensity of\nBlautia A wexlerae (arbitrary unit)") + xlab('')+ggtitle("16S-PC")

```


```{r}

Fig1B = p4 + p5

```



## Figure 1.C: Relative abundance of significant features between ASD and NT groups in MTG and MTT


```{r}

#### MTG: K07493

PS_kegg_fun_MTG_age_filtered_paired_family <- readRDS("G:/Shared drives/M3-shared/Data/MTG/Step3_Making_phyloseq/age_filtered_PS_objects/PS_kegg_fun_MTG_age_filtered_paired_family.rds")

PS_kegg_fun_MTG_age_filtered_paired_family <- sg_normalize(PS_kegg_fun_MTG_age_filtered_paired_family, method = "tss", log = FALSE)


PS_kegg_fun_MTG_age_filtered_paired_family  <- prune_taxa(taxa_names(PS_kegg_fun_MTG_age_filtered_paired_family) %in% c('K07493'), PS_kegg_fun_MTG_age_filtered_paired_family)


plot_MTG <- psmelt(PS_kegg_fun_MTG_age_filtered_paired_family)


plot_MTG <- as.data.table(plot_MTG)

          # add very small value, min/100000 to 0

          plot_MTG$Abundance <- plot_MTG$Abundance+min(plot_MTG[Abundance!=0]$Abundance)/100000

          

# relabel

plot_MTG[, Host.disease.status:=ifelse(Host.disease.status=="Case (disease status)", "ASD", "NT")]

          

p6 <- ggplot(data=plot_MTG, aes_string(x='Host.disease.status', y = 'Abundance', fill='Host.disease.status')) +

            geom_boxplot(outlier.color=NA) +

            geom_jitter(position=position_jitterdodge(0.2), color="gray44") + scale_y_log10() + 

            scale_fill_manual(values=c("#84CF04","#01B5BB"))+

            theme_bw() +

            theme(text = element_text(size=10), legend.position="none") +ylab("Relative abundance of\nK07493") + xlab('')+ggtitle("MTG-KEGG")

p6
```



```{r}

### MTT Biocyc


PS_biocyc_MTT_age_filtered <- readRDS("/Volumes/GoogleDrive/Shared drives/M3-shared/Data/MTT/Step3_Making_phyloseq/age_filtered_PS_objects/PS_biocyc_MTT_age_filtered.rds")


PS_biocyc_MTT_age_filtered <- sg_normalize(PS_biocyc_MTT_age_filtered, method = "tss", log = FALSE)


PS_biocyc_MTT_age_filtered  <- prune_taxa(taxa_names(PS_biocyc_MTT_age_filtered) %in% c('RXN-16648'), PS_biocyc_MTT_age_filtered )


plot_MTT <- psmelt(PS_biocyc_MTT_age_filtered)


plot_MTT <- as.data.table(plot_MTT)

          # add very small value, min/100000 to 0

          plot_MTT$Abundance <- plot_MTT$Abundance+min(plot_MTT[Abundance!=0]$Abundance)/100000

          

# relabel

plot_MTT[, Host.disease.status:=ifelse(Host.disease.status=="Case (disease status)", "ASD", "NT")]


p7 <- ggplot(data=plot_MTT, aes_string(x='Host.disease.status', y = 'Abundance', fill='Host.disease.status')) +

            geom_boxplot(outlier.color=NA) + 

            geom_jitter(position=position_jitterdodge(0.2), color="gray44") + scale_y_log10() + 

            scale_fill_manual(values=c("#84CF04","#01B5BB"))+

            theme_bw() +

            theme(text = element_text(size=10), legend.position="none") +ylab("Relative abundance of\nRXN-16648") + xlab('')+ggtitle("MTT-BioCyc")


```


## Figure 1.D. Measured intensity (log transformed) of 5-dodecenoate between the ASD and NT groups


```{r pressure, echo=FALSE}

MTB <- readRDS('/Volumes/GoogleDrive/Shared drives/M3-shared/Data/Metabolomics/INRD17_0415_metab_ps_age_filtered_prevalence_filtered_zero_imputed.rds')

otu_table(MTB) <- log2(otu_table(MTB))


MTB <- prune_taxa(taxa_names(MTB) == 'ch__23584', MTB)


plot <- psmelt(MTB)


plot <- data.table(plot)

# relabel

plot[, Host.disease.status:=ifelse(Host.disease.status=="Case (disease status)", "ASD", "NT")]


p8 <- ggplot(data=plot, aes_string(x='Host.disease.status', y = 'Abundance', fill='Host.disease.status')) +

            geom_boxplot(outlier.color=NA) + 

             geom_jitter(position=position_jitterdodge(0.2), color="gray44") + scale_y_log10() + 

            scale_fill_manual(values=c("#84CF04","#01B5BB"))+

            theme_bw() +

            theme(text = element_text(size=10), legend.position="none") +ylab("Measured intensity\n(log transformed) of 5-dodecenoate") + xlab('')+ggtitle("MTB")

```


## Figure 1.E. Significant Spearman correlation around 5-dodecenoate (FDR <0.05 and Rho > |0.3|) between each pair of measurements


```{r, fig.height = 10, fig.width=15}

# I used here the filtered data, we only keep interesting features for the analysis (see nodes_dodecenoate_analysis sheet to have the detail of the analysis) . Otherwise the network looked too crowded. For reference, the complete files are in the same folder (nodes_dodecenoate_complete and edges_dodecenoate_complete). 


Nodes_dodecenoate <- read.csv('/Volumes/GoogleDrive/Shared drives/M3-shared/publications/Omics_integration/data_for_the_ms/Microbiome_Metabolites/Network/nodes_dodecenoate_filtered.csv')

Edges_dodecenoate <- read.csv('/Volumes/GoogleDrive/Shared drives/M3-shared/publications/Omics_integration/data_for_the_ms/Microbiome_Metabolites/Network/edges_dodecenoate_filtered.csv')


# relabel

Nodes_dodecenoate$Dataset <- ifelse(Nodes_dodecenoate$Dataset=="MTG_sourdough", "MTG-Sour",

                                    ifelse(Nodes_dodecenoate$Dataset=="16SV4", "16S-V4",

                                           ifelse(Nodes_dodecenoate$Dataset=="MTG_biocyc", "MTG-BioCyc",

                                                  ifelse(Nodes_dodecenoate$Dataset=="MTT_biocyc", "MTT-BioCyc", 

                                                         ifelse(Nodes_dodecenoate$Dataset=="MTB", "MTB", "Other")))))

setnames(Edges_dodecenoate, "status", "Status")


Network <- tbl_graph(nodes = Nodes_dodecenoate, edges = Edges_dodecenoate, directed = FALSE)


p9 <- ggraph(Network, layout = "nicely")  +geom_edge_link(aes(color = Status)) +

  geom_node_point(aes(shape = Dataset, color = Species), size = 3)+

  geom_node_text(aes(label = label), repel = TRUE, size = 3)  + scale_shape_manual(values=c(15, 16, 17, 18, 10)) +

  theme_graph() +

            theme(text = element_text(size=5), legend.title = element_text(size = 10),

  legend.text = element_text(size = 10)) 


```


## Fig 1 in one plot command

```{r fig.height=10, fig.width=15}

plot_grid(plot_grid(plot_grid(p1, p2, p3, labels=NULL, nrow=1, rel_widths=c(2, 1, 1)),    # 1a

                    plot_grid(p4, p5, labels=NULL, nrow=1, rel_widths=c(1.5, 1)),    # 1b

                    plot_grid(p6, p7,  p8, labels=c("c", "d", "e"), nrow=1),    # 1c

                    labels=c("a", "b", ""), ncol=1),    # left panel 

          NULL, p9, labels=c("", "", "f"), nrow=1,  rel_widths = c(1, 0.2, 2.1))    # adding empty panel in the middle to make some space

```


# Fig. 2 - HALLA

This figure is a set of direct output from HALLA software, edited by Inkscape.  

Original figures are in /Volumes/GoogleDrive/Shared drives/M3-shared/publications/Omics_integration/data_for_the_ms/Microbiome_Metabolites/HALLA/Outputs/  

and Inkscape version is, /Volumes/GoogleDrive/Shared drives/M3-shared/publications/Omics_integration/Figures/Figure2_new.svg


# Fig. 3 - MelonnPan prediction MTMA

```{r fig.width=5, fig.height=4}

melonn_mtma_res <- fread("/Volumes/GoogleDrive/Shared drives/M3-internal/Metabolites_Nomination/MelonnPan/MelonnPan_MTMA/KEGG_BioCyc_RMA_MV_combined.csv")


# add chemname

melonn_mtma_res$chemName <- as.character(lapply(strsplit(as.character(melonn_mtma_res$chemID_all), split="\\|"), "[", 2))


# reformat for plot

melonn_mtma_plot <- melt.data.table(melonn_mtma_res, id.vars=c("sg_chemID", "chemID_all", "chemName")) %>%

  .[, DB:=ifelse(grepl("BioCyc", variable), "BioCyc", "KEGG")] %>%

  .[, type:=ifelse(grepl("Effect_size", variable), "Effect_size", 

                   ifelse(grepl("Pvalue", variable), "Pvalue", "Other"))] %>%

  .[type %in% c("Effect_size", "Pvalue"), list(sg_chemID, chemID_all, chemName, value, DB, type)] %>%

  dcast.data.table(., sg_chemID+chemID_all+chemName+DB ~ type, value.var="value") %>%

  .[Pvalue<0.05, chem_name:=chemName] %>%

  .[, Significance:=ifelse(Pvalue<0.05, "p-value < 0.05", "p-value>0.05")]


# only BioCyc

ggplot(data=melonn_mtma_plot[DB=="BioCyc"], aes(x=Effect_size, y=-log10(Pvalue), label=chem_name, colour=Significance))+

  geom_point()+

  geom_hline(yintercept=(-1)*log10(0.05), linetype=2, colour="gray")+

  geom_vline(xintercept=0, linetype=2, colour="gray")+

  geom_text_repel()+ 

  xlab("log2 fold change, ASD over NT")+

  ylim(c(0, 2))+

  scale_colour_manual(values=c("black","gray"))+

 # geom_text(aes(x=-0.43, y=(-1)*log10(0.05), label="p-value=0.05", vjust=-0.7))+    # add p-value=0.05 label

  theme_bw()+

  theme(legend.position="none")

```





# Figures below are related to MARA score - decided not to include in the paper (all chunks are eval=FALSE)


# Figure 2: Focus on ASD severity and the microbiome using MARA score as contrast 


## Figure 2.A. Alpha diversity


### 16S V4


```{r eval=FALSE}

alpha_16S <- read.csv('/Volumes/GoogleDrive/Shared drives/M3-shared/publications/Omics_integration/data_for_the_ms/16S V4/All time points/MARA/16SV4_AllTP_MARA__Alphadiversity_2020-07-25.csv')


ggplot(alpha_16S, aes_string(x='Mobile.Autism.Risk.Assessment.Score', y='value', colour='Mobile.Autism.Risk.Assessment.Score')) + geom_point(size=4, shape = 15) + facet_wrap(~variable, scales = "free", ncol = 3) + theme_bw() +theme(legend.position = "none", axis.text=element_text(size=12), axis.title = element_text(size = 20)) + labs(y='Alpha diversity measure', x = 'MARA score')

```

### 16S PC


```{r eval=FALSE}

alpha_16SPC <- read.csv('/Volumes/GoogleDrive/Shared drives/M3-shared/publications/Omics_integration/data_for_the_ms/16S PC/Strain approach/MARA/StrainApproach_MARA__16S PC_Alphadiversity_2020-05-22.csv')


# I want the same width than the other alpha diversity plot but I couldn't find a way to adjust. I used facet_wrap with 2 other plots. We will keep only the richness here

alpha_16S <- alpha_16S %>% filter(variable != 'Richness')


alpha_16SPC_plot <- full_join(alpha_16SPC, alpha_16S)


ggplot(alpha_16SPC_plot, aes_string(x='Mobile.Autism.Risk.Assessment.Score', y='value', colour='Mobile.Autism.Risk.Assessment.Score')) + geom_point(size=5, shape = 8) + facet_wrap(~variable, scales = "free", ncol = 3) + theme_bw() +theme(legend.position = "none", axis.text=element_text(size=12), axis.title = element_text(size = 20)) + labs(y='Alpha diversity measure', x = 'MARA score')

```


## Figure 2.B - Akkermansiaceae


### 16S V4 	


```{r eval=FALSE}

stat_family <- read.csv('/Volumes/GoogleDrive/Shared drives/M3-shared/publications/Omics_integration/data_for_the_ms/16S V4/All time points/MARA/16SV4_AllTP_MARA__Composition_Family2020-07-25.csv')


# Same here in order to have the same width I had 2 species that are not relevant so we won't use them.

stat_family <- stat_family %>% filter(Family == c('Akkermansiaceae','Lachnospiraceae','Bacteroidaceae'))

composition <- ggplot(stat_family, aes_string(x='Mobile.Autism.Risk.Assessment.Score', y='Abundance', colour='Mobile.Autism.Risk.Assessment.Score')) + geom_point(size=4, shape =15) + facet_wrap(~Family, scale = 'free') + theme_bw() + theme(axis.title.x = element_blank(),axis.text=element_text(size=12), axis.title = element_text(size = 20), legend.position = "none") + labs(y='Abundance', x ='Mobile.Autism.Risk.Assessment.Score') + scale_y_continuous(limits = c(0, 0.08)) 

composition



```


### 16S PC


```{r eval=FALSE}

PC <- read.csv('/Volumes/GoogleDrive/Shared drives/M3-shared/publications/Omics_integration/data_for_the_ms/16S PC/Strain approach/MARA/StrainApproach_MARA__16S PC_Composition_Family2020-05-22.csv')


PC <- PC %>% filter(Family == 'Akkermansiaceae'|Family == 'Lachnospiraceae'|Family == 'Bacteroidaceae')


ggplot(PC, aes_string(x='Mobile.Autism.Risk.Assessment.Score', y='Abundance', colour='Mobile.Autism.Risk.Assessment.Score')) + geom_point(size=5, shape = 8) + facet_wrap(~Family, scales = "free", ncol = 3) + theme_bw() +theme(axis.title.x = element_blank(), legend.position = "none", axis.text=element_text(size=12), axis.title.y = element_text(size = 20)) + labs(y='Relative abundance')

```


# Figure 3: in this figure we are comapring the most severe cases of ASD to their NT siblings 


## Figure 3.A: Distribution of ASD severity; plot generated by Ellie from a previous analysis CNS_56_M3_MARA


```{r eval=FALSE}

meta_asd <- read.csv('/Volumes/GoogleDrive/Shared drives/M3-shared/publications/Omics_integration/data_for_the_ms/Metadata/ASD_metadata.csv')


meta_asd$Mobile.Autism.Risk.Assessment.Score..Biospecimen. <- as.numeric(as.character(meta_asd$Mobile.Autism.Risk.Assessment.Score..Biospecimen.))


# Sumamry of MARA

summary(meta_asd$Mobile.Autism.Risk.Assessment.Score..Biospecimen.)


# Subjects number per bin

meta_asd$range <- cut(meta_asd$Mobile.Autism.Risk.Assessment.Score..Biospecimen., breaks = -10:10)


table(meta_asd$range)


# Plot distribution

ggplot(meta_asd, aes(x=Mobile.Autism.Risk.Assessment.Score..Biospecimen.)) + geom_histogram(binwidth = 0.8)  + ylab("Number of subjects") + xlab("MARA score") + scale_x_continuous(breaks = seq(-10, 4, 2), lim = c(-10, 4)) + scale_y_continuous(breaks = seq(0, 27, 4), lim = c(0, 27)) + theme_bw() + theme(text = element_text(size=20), legend.title = element_text(size = 30), legend.text = element_text(size = 20)) 



```



## Figure 3.B: Depletion of ASV876


```{r eval=FALSE}

ps_not_norm_pass_min_postDD_age_filtered <- readRDS("/Volumes/GoogleDrive/Shared drives/M3-shared/Data/V4/200312_ASVdata8_updateAsteria/Phyloseq_objects/age_filtered_PS_objects/ps_not_norm_pass_min_postDD_age_filtered.rds")

# tss normaliszation 

ps_not_norm_pass_min_postDD_age_filtered <- sg_normalize(ps_not_norm_pass_min_postDD_age_filtered , method = "tss", log = FALSE)

# MARA score was not numeric 

ps_not_norm_pass_min_postDD_age_filtered@sam_data[['Mobile.Autism.Risk.Assessment.Score']] <- as.numeric(as.character(ps_not_norm_pass_min_postDD_age_filtered@sam_data[['Mobile.Autism.Risk.Assessment.Score']]))


# Keeping only the family ID of the lowest MARA score

ASD = subset_samples(ps_not_norm_pass_min_postDD_age_filtered, ps_not_norm_pass_min_postDD_age_filtered@sam_data[['Mobile.Autism.Risk.Assessment.Score']] < -8)

family_id <- ASD@sam_data[['Family.group.ID']] 


ps_not_norm_pass_min_postDD_age_filtered <- subset_samples(ps_not_norm_pass_min_postDD_age_filtered, ps_not_norm_pass_min_postDD_age_filtered@sam_data[['Family.group.ID']] %in% family_id)


# ASV 876 is Faecalicatena lactaris

ASV876 <- 'GCAAGCGTTATCCGGATTTACTGGGTGTAAAGGGAGCGTAGACGGAGCAGCAAGTCTGATGTGAAAACCCGGGGCTCAACCCCGGGACTGCATTGGAAACTGTTGATCTGGAGTGCCGGAGAGGTAAGCGGAATTCCTAGTGTAGCGGTGAAATGCGTAGATATTAGGAGGAACACCAGTGGCGAAGGCGGCTTACTGGACGGTAACTGACGTTGAGGCTCGAAAGCGTGGGG' 


keep <- c(ASV876)

ps_not_norm_pass_min_postDD_age_filtered  <- prune_taxa(taxa_names(ps_not_norm_pass_min_postDD_age_filtered) %in% keep, ps_not_norm_pass_min_postDD_age_filtered)


plot <- psmelt(ps_not_norm_pass_min_postDD_age_filtered)

          

plot <- as.data.table(plot)

          # add very small value, min/100000 to 0

          plot$Abundance <- plot$Abundance+min(plot[Abundance!=0]$Abundance)/100000

          

          ggplot(data=plot, aes_string(x='Within.study.sampling.date', y = 'Abundance', fill='Host.disease.status')) +

            geom_boxplot(outlier.color=NA) + 

            geom_jitter(position=position_jitterdodge(0.2), cex=4, color="gray44", shape = 15) + scale_y_log10() + 

            scale_fill_manual(values=c("#84CF04","#01B5BB"))+

            theme_bw() +

            theme(text = element_text(size=20), axis.text.x = element_text(angle = 45, vjust = 1, hjust=1), legend.title = element_blank(),

  legend.text = element_blank()) +ylab("Abundance of ASV876") 


```


## Figure 3.C: Volcano plot after feature selection in MTG


```{r eval=FALSE}

library(ggrepel)

Feature_selection_MTG <- read.csv('/Volumes/GoogleDrive/Shared drives/M3-shared/publications/Omics_integration/data_for_the_ms/Metagenomics/Kegg- Function/HDS_lowest_MARA/MTG_KEGG_LM__Feature_Selection_2020-08-14 12.15.25 PM.csv')


Feature_selection_MTG <- Feature_selection_MTG %>%  mutate(threshold = case_when(

  padj<0.05 ~ "0.05",

  padj<0.1 ~ "0.1",

  padj<0.2 ~ "0.2",

  TRUE ~ 'none'

)) %>% mutate(color = case_when(

  threshold==0.05 ~ "#84CF04",

  threshold==0.1 ~ "#01B5BB",

  threshold==0.2 ~ "#E50E63",

  threshold=="none" ~ "#808080")) %>% mutate(label = case_when(

    (padj<0.05)&(log2FoldChange < -1)&(Definition != 'uncharacterized protein') ~ Gene,

    TRUE ~ ''

  ))



Feature_selection_MTG$threshold <- factor(Feature_selection_MTG$threshold, levels = c("0.05","0.1","0.2", "none"))

Feature_selection_MTG <- Feature_selection_MTG[order(Feature_selection_MTG$threshold), ]



ggplot(Feature_selection_MTG, aes(x=log2FoldChange, y=log10(pvalue)*-1, color=threshold, label=label))  + geom_text_repel(size=4,color="black",force = 8) + geom_point(size=3,  shape = 4)+ylab("-log10(p-value)") + scale_colour_manual(values = unique(Feature_selection_MTG$color)) + xlab("log-2 fold-change of ASD/NT") +

  geom_hline(aes(yintercept=log10(0.05)*-1, linetype="pvalue= 0.05"), color="red")  +  geom_hline(aes(yintercept=log10(0.1)*-1, linetype="pvalue= 0.1"), color="red") + geom_hline(aes(yintercept=log10(0.01)*-1, linetype="pvalue= 0.01"), color="red") +   guides(color=guide_legend(title="adjusted pvalue")) + scale_linetype_manual(name = "guide", values = c("twodash", "dotted", "longdash"), guide = guide_legend(override.aes = list(color = "red"))) + theme_bw() + theme( axis.text=element_text(size=16), axis.title = element_text(size = 20), legend.title = element_text(size = 15),

  legend.text = element_text(size = 15)

  ) 

```


## Figure 3.D: MTB


```{r, echo=FALSE, eval=FALSE}


physeq_object <- readRDS('~/Google Drive File Stream/Shared drives/M3-internal/Metabolites_Nomination/Measured/Measured_FS/INRD17_0415_metab_ps_age_filtered_prevalence_filtered_zero_imputed.rds')

otu_table(physeq_object) <- log2(otu_table(physeq_object))


physeq_object@sam_data[['Mobile.Autism.Risk.Assessment.Score']] <- as.numeric(as.character(physeq_object@sam_data[['Mobile.Autism.Risk.Assessment.Score']]))



ASD = subset_samples(physeq_object, physeq_object@sam_data[['Mobile.Autism.Risk.Assessment.Score']] < -8)

family_id <- ASD@sam_data[['Family.group.ID']] #Keeping only the family ID of the lowest MARA score


physeq_object <- subset_samples(physeq_object, physeq_object@sam_data[['Family.group.ID']] %in% family_id)



physeq_object <- prune_taxa(taxa_names(physeq_object) == 'ch__23701', physeq_object) # keep only CEGABA


plot <- psmelt(physeq_object)




          ggplot(data=plot, aes_string(x='Host.disease.status', y = 'Abundance', fill='Host.disease.status')) +

            geom_boxplot(outlier.color=NA) + 

            geom_jitter(position=position_jitterdodge(0.2), cex=4, color="gray44", shape = 16) + 

            scale_fill_manual(values=c("#84CF04","#01B5BB"))+

            theme_bw() +

            theme(axis.title.x = element_blank(),axis.text=element_text(size=16), text = element_text(size=20), legend.title = element_text(size = 10),

  legend.text = element_text(size = 10)) +ylab('Measured intensity of CEGABA') 

```


## Figure 3.E. Network analysis. Measured intensity (log transformed) of CEGABA between the ASD and NT groups

E.Significant Spearman correlation around CEGABA (FDR <0.05 and Rho > |0.3|) between each pair of measurements


```{r fig.height = 10, fig.width=15, eval=FALSE}

# I used here the filtered data, we only keep interesting features for the analysis (see nodes_CEGABA_analysis sheet to have the detail of the analysis). Otherwise the network looked too crowded. For reference, the complete files are in the same folder (nodes_CEGABA_complete and edges_CEGABA_complete). 


Nodes_CEGABA <- read.csv('/Volumes/GoogleDrive/Shared drives/M3-shared/publications/Omics_integration/data_for_the_ms/Microbiome_Metabolites/Network/nodes_CEGABA_filtered.csv')

Edges_CEGABA <- read.csv('/Volumes/GoogleDrive/Shared drives/M3-shared/publications/Omics_integration/data_for_the_ms/Microbiome_Metabolites/Network/edges_CEGABA_filtered.csv')


Network_CEGABA <- tbl_graph(nodes = Nodes_CEGABA, edges = Edges_CEGABA, directed = FALSE)


ggraph(Network_CEGABA, layout = "nicely")  +geom_edge_link(aes(color = status)) +

  geom_node_point(aes(shape = Dataset, color = Species), size = 9)+

  geom_node_text(aes(label = label), repel = TRUE, size = 6.5)  + scale_shape_manual(values=c(15, 16, 17, 18, 10)) +

  theme_graph() +

            theme(text = element_text(size=15), legend.title = element_text(size = 15),

  legend.text = element_text(size = 15)) 


```


# Figure 5: Mice experiments


## Figure 5.A: Volcano plot on mouse brain gene expression between 5-dodecanoate feeding and sham controls


```{r eval=FALSE}

FS_Dodecenoic_acid <- readRDS("~/Downloads/FS_Dodecenoic_acid.rds")

FS <- FS_Dodecenoic_acid$fs_all


Results <- FS$results

Results <- Results %>%  mutate(threshold = case_when(

  pvalue<0.05 ~ "0.05",

  pvalue<0.1 ~ "0.1",

  pvalue<0.2 ~ "0.2",

  TRUE ~ 'none'

)) %>% mutate(color = case_when(

  threshold==0.05 ~ "#84CF04",

  threshold==0.1 ~ "#01B5BB",

  threshold==0.2 ~ "#E50E63",

  threshold=="none" ~ "#808080"))



Results$threshold <- factor(Results$threshold, levels = c("0.05","0.1","0.2", "none"))

Results <- Results[order(Results$threshold), ]



ggplot(Results, aes(x=log2FoldChange, y=log10(pvalue)*-1, color=threshold))  + geom_point(size=2) +ylab("-log10(p-value)") + scale_colour_manual(values = unique(Feature_selection_MTG$color)) + xlab("log-2 fold-change") +

  geom_hline(aes(yintercept=log10(0.05)*-1, linetype="pvalue= 0.05"), color="red")  +  geom_hline(aes(yintercept=log10(0.1)*-1, linetype="pvalue= 0.1"), color="red") + geom_hline(aes(yintercept=log10(0.2)*-1, linetype="pvalue= 0.01"), color="red") +   guides(color=guide_legend(title="pvalue")) + scale_linetype_manual(name = "guide", values = c("twodash", "dotted", "longdash"), guide = guide_legend(override.aes = list(color = "red"))) + theme_bw() + theme( axis.text=element_text(size=16), axis.title = element_text(size = 20), legend.title = element_text(size = 15),

  legend.text = element_text(size = 15)

  ) 

```


## Figure 5.B: Top 10 pathways identified by GSEA. Node size is related to gene numbers in each pathway and edges indicates genes shared by multiple pathways. 

Ellie generated this figure in the following HTML: OMICS310_INRD20_0523_Step4_Pathway_GSEA_2020-10-18.html in the shared drive



# Tables


## Table 1. M3 study cohort characteristics 

```{r}

meta <- import_metadata_from_kb(Sys.getenv("kb_username"),Sys.getenv("kb_password"), report_collection_id=160)

#meta<- fread("/Volumes/GoogleDrive/Shared drives/M3-shared/Data/V4/00_Metadata/INRD17_0415 (M3) Biospecimen 200312.csv")


meta_age_filter <- meta %>% filter(Age..months...Biospecimen. >23) # 116 families

meta_unique <- meta_age_filter %>% distinct(Host.ID, .keep_all = TRUE)


meta_age_filter_ASD <- meta_age_filter %>% filter(Host.disease.status..Biospecimen. == 'Case (disease status)') %>% distinct(Host.ID, .keep_all = TRUE)


meta_age_filter_NT <- meta_age_filter %>% filter(Host.disease.status..Biospecimen. == 'Control (disease status)') %>% distinct(Host.ID, .keep_all = TRUE)



#meta_age_filter_NT <- meta_age_filter_NT %>% distinct(Host.ID, .keep_all = TRUE)


```


```{r Age}

mean_age_ASD <- mean(as.numeric(meta_age_filter_ASD$Age..months...Biospecimen.))

sd_age_ASD <- sd(as.numeric(meta_age_filter_ASD$Age..months...Biospecimen.))


mean_age_NT <- mean(as.numeric(meta_age_filter_NT$Age..months...Biospecimen.))

sd_age_NT <- sd(as.numeric(meta_age_filter_NT$Age..months...Biospecimen.))


Age_pval <- t.test(as.numeric(meta_age_filter_ASD$Age..months...Biospecimen.), as.numeric(meta_age_filter_NT$Age..months...Biospecimen.))


```


```{r Gender}

male_ASD <- meta_age_filter_ASD %>% filter(meta_age_filter_ASD$Biological.sex..Biospecimen. == 'Male')

male_ASD_percentage <- length(male_ASD$Host.ID)/length(meta_age_filter_ASD$Host.ID)


male_NT <- meta_age_filter_NT %>% filter(meta_age_filter_NT$Biological.sex..Biospecimen. == 'Male')

male_NT_percentage <- length(male_NT$Host.ID)/length(meta_age_filter_NT$Host.ID)

tbl <- table(meta_unique$Biological.sex..Biospecimen., meta_unique$Host.disease.status..Biospecimen.)


Male_pval <- chisq.test(tbl)


```


```{r Functional bowel findings}

# Removing the NAs: 

meta_age_filter_ASD_NA <- meta_age_filter_ASD %>% filter(!is.na(Functional.bowel.finding..Biospecimen.))


ASD_constipation <- meta_age_filter_ASD_NA %>% filter(Functional.bowel.finding..Biospecimen. == 'Tends to have constipation')

ASD_constipation_percentage <- length(ASD_constipation$Host.ID)/length(meta_age_filter_ASD_NA$Host.ID)


ASD_normal <- meta_age_filter_ASD_NA %>% filter(Functional.bowel.finding..Biospecimen. == 'Tends to have normal bowel function')

ASD_normal_percentage <- length(ASD_normal$Host.ID)/length(meta_age_filter_ASD_NA$Host.ID)


ASD_diarrhea <- meta_age_filter_ASD_NA %>% filter(Functional.bowel.finding..Biospecimen. == 'Tends to have diarrhea')

ASD_diarrhea_percentage <- length(ASD_diarrhea$Host.ID)/length(meta_age_filter_ASD_NA$Host.ID)


# NT


meta_age_filter_NT_NA <- meta_age_filter_NT %>% filter(!is.na(Functional.bowel.finding..Biospecimen.))


NT_constipation <- meta_age_filter_NT_NA %>% filter(Functional.bowel.finding..Biospecimen. == 'Tends to have constipation')

NT_constipation_percentage <- length(NT_constipation$Host.ID)/length(meta_age_filter_NT_NA$Host.ID)


NT_normal <- meta_age_filter_NT_NA %>% filter(Functional.bowel.finding..Biospecimen. == 'Tends to have normal bowel function')

NT_normal_percentage <- length(NT_normal$Host.ID)/length(meta_age_filter_NT_NA$Host.ID)


NT_diarrhea <- meta_age_filter_NT_NA %>% filter(Functional.bowel.finding..Biospecimen. == 'Tends to have diarrhea')

NT_diarrhea_percentage <- length(NT_diarrhea$Host.ID)/length(meta_age_filter_NT_NA$Host.ID)


Constipation <- meta_unique %>% filter(Functional.bowel.finding..Biospecimen. == 'Tends to have constipation')


tbl_constipation <- table(Constipation$Functional.bowel.finding..Biospecimen., Constipation$Host.disease.status..Biospecimen.)

Constipation_pval <- chisq.test(tbl_constipation)


Normal <- meta_unique %>% filter(Functional.bowel.finding..Biospecimen. == 'Tends to have normal bowel function')


tbl_normal <- table(Normal$Functional.bowel.finding..Biospecimen., Normal$Host.disease.status..Biospecimen.)

Normal_pval <- chisq.test(tbl_normal)


diarrhea <- meta_unique %>% filter(Functional.bowel.finding..Biospecimen. == 'Tends to have diarrhea')


tbl_diarrhea <- table(diarrhea$Functional.bowel.finding..Biospecimen., diarrhea$Host.disease.status..Biospecimen.)

diarrhea_pval <- chisq.test(tbl_diarrhea)

```


```{r MARA}

mean_MARA <- mean(as.numeric(meta_age_filter_ASD$Mobile.Autism.Risk.Assessment.Score..Biospecimen.))

sd_MARA <- sd(as.numeric(meta_age_filter_ASD$Mobile.Autism.Risk.Assessment.Score..Biospecimen.))

```


```{r}

sessionInfo()

```