# Libraries
```{r}
library("rstan")
library("reshape2")
library("stringr")
library("phyloseq")
#library("ggscaffold")
library("feather")
library(ggplot2)
library(ggpubr)
library(vegan)

library(DirichletMultinomial)
library(lattice)
library(xtable)
library(parallel)
library(pheatmap)

library(tm)
library(slam)
library(topicmodels)
library(Boruta)


library(dplyr)
library(colorspace)
source("lda_helper_functions.R")
```

# Read in data
```{r}

ps_mtg <- readRDS("../data/mtg/ps_rle_nooutliers.rds")
ps_mtg@sam_data$familyID <- as.character(ps_mtg@sam_data$familyID)
ps_16s <- readRDS("../results/topic_modeling/ps_16s_dds_taxannotation.rds")
ps_mtt <- readRDS("../data/mtt/ps_mtt_rle_nooutliers.rds")
ps_mbx <- readRDS("../data/metabol/ps_mbx_rle_nooutliers_fixedmapping.rds")
```



# Rarefy MTT samples with abnormally high counts
```{r}
data <- t(ps_mtg@otu_table)
hist(apply(data, 1, max), main = "original distribution")

thresh <- mean(apply(data, 1, max)) + sd(apply(data, 1, max))/sqrt(nrow(data))
keep <- apply(data, 1, max) < thresh
print(sum(keep))
rarefy_samples <- rrarefy(round(data[!keep, ]), sample = round(thresh))

data <- rbind(data[keep, ], rarefy_samples)
hist(apply(data, 1, max), main = "new distribution", breaks= 200)

ps_mtg_rarefy_outliers <- phyloseq(otu_table(t(data), taxa_are_rows = T),sample_data(ps_mtg@sam_data),tax_table(ps_mtg@tax_table))
#ps_mtg_rarefy_outliers <- prune_samples(sample_names(ps_mtg_rarefy_outliers) != "M3_177_P1_A_V1", ps_mtg_rarefy_outliers)

ord <- ordinate(ps_mtg_rarefy_outliers)
plot_ordination(ps_mtg_rarefy_outliers, ord, label = "biospecimen_name")

saveRDS(ps_mtg_rarefy_outliers, file = "../data/mtg/ps_mtg_rle_nooutliers_fixedmapping_nooutliers.rds")
```



# Rarefy MTT samples with abnormally high counts
```{r}
data <- t(ps_mtt@otu_table)
hist(apply(data, 1, max), main = "original distribution")

thresh <- mean(apply(data, 1, max)) + sd(apply(data, 1, max))/sqrt(nrow(data))
keep <- apply(data, 1, max) < thresh
print(sum(keep))
rarefy_samples <- rrarefy(round(data[!keep, ]), sample = round(thresh))

data <- rbind(data[keep, ], rarefy_samples)
hist(apply(data, 1, max),main = "new distribution")
ps_mtt_rarefy_outliers <- phyloseq(otu_table(t(data), taxa_are_rows = T),sample_data(ps_mtt@sam_data), tax_table(ps_mtt@tax_table))
#ps_mtt_rarefy_outliers <- prune_samples(sample_names(ps_mtt_rarefy_outliers) != "M3_177_P1_A_V1", ps_mtt_rarefy_outliers)

ord <- ordinate(ps_mtt_rarefy_outliers)
plot_ordination(ps_mtt_rarefy_outliers, ord, label = "biospecimen_name")

saveRDS(ps_mtt_rarefy_outliers, file = "../data/mtt/ps_mtt_rle_nooutliers_fixedmapping_nooutliers.rds")
```


# Rarefy MBX samples with abnormally high counts in 1 metabolite
```{r}
data <- t(ps_mbx@otu_table)
hist(apply(data, 1, max), main = "Each sample's max count for single metabolite")
hist(apply(data, 2, max), main = "Each metabolites's max count for single sample")
hist(rowSums(data), breaks=20)

thresh <- mean(apply(data, 1, max)) + sd(apply(data, 1, max))/sqrt(nrow(data))
keep <- apply(data, 1, max) < thresh
print(sum(keep))
hist(apply(data[keep, ], 1, max))
rarefy_samples <- rrarefy(round(data[!keep, ]), sample = round(thresh))

data <- rbind(data[keep, ], rarefy_samples)
hist(apply(data[keep, ], 1, max))
ps_mbx_rarefy_outliers <- phyloseq(otu_table(t(data), taxa_are_rows = T),sample_data(ps_mbx@sam_data),tax_table(ps_mbx@tax_table))
ps_mbx_rarefy_outliers <- prune_samples(sample_names(ps_mbx_rarefy_outliers) != "M3_177_P1_A_V1", ps_mbx_rarefy_outliers)

ord <- ordinate(ps_mbx_rarefy_outliers)
plot_ordination(ps_mbx_rarefy_outliers, ord, label = "biospecimen_name")

saveRDS(ps_mbx_rarefy_outliers, file = "../data/metabol/ps_mbx_rle_nooutliers_fixedmapping_nooutliers.rds")
```

# Max likelihood number of topics
```{r}

p_16s <- plotNumberTopics(directory = "../results/latent_variable_modeling/16s_models/", title = "16s")
ggsave("../results/latent_variable_modeling/model_selection/likelihood_16s.pdf", p_16s, height = 2, width = 3.5)

p_mtg <- plotNumberTopics(directory = "../results/latent_variable_modeling/mtg_models/", title = "MTG")
ggsave("../results/latent_variable_modeling/model_selection/likelihood_mtg.pdf", p_mtg, height = 2, width = 3.5)

p_mtt <- plotNumberTopics(directory = "../results/latent_variable_modeling/mtt_models/", title = "MTT")
ggsave("../results/latent_variable_modeling/model_selection/likelihood_mtt.pdf", p_mtg, height = 2, width = 3.5)

p_mbx <- plotNumberTopics(directory = "../results/latent_variable_modeling/mbx_models/", title = "MBX")
ggsave("../results/latent_variable_modeling/model_selection/likelihood_mbx.pdf", p_mbx, height = 2, width = 3.5)

```

# Read in data
```{r}
ps_mtg <- readRDS("../../data/mtg/ps_mtg_rle_nooutliers_adjcounts.rds")
ps_mtg@sam_data$familyID <- as.character(ps_mtg@sam_data$familyID)
ps_16s <- readRDS("../../results/topic_modeling/ps_16s_dds_taxannotation.rds")
ps_mtt <- readRDS("../../data/mtt/ps_mtt_rle_nooutliers_adjcounts.rds")
ps_mbx <- readRDS("../../data/mbx/ps_mbx_rle_nooutliers_adjcounts_fixedmapping.rds")
```


# Filter
```{r}
ps_16s_filt <- filterByPrevalence(ps_16s, 0.05)
ps_mtg_filt <- filterByPrevalence(ps_mtg, 0.1)
ps_mtt_filt <- filterByPrevalence(ps_mtt, 0.1)
ps_mbx_filt <- filterByPrevalence(ps_mbx, 0.1)
```

# Simulate counts
```{r}
model_16s <- readRDS("../results/latent_variable_modeling/16s_models/model_16s_10topics_10000iter_0.05filter.rds")
sim_counts_16s <- getSimulatedDataset(ps_16s_filt, model_16s, n = "sample_counts")
saveRDS(sim_counts_16s, "sim_counts_16s_10topics.rds")

model_mtg <- readRDS("../results/latent_variable_modeling/mtg_models/model_mtg_2topics_3000iter_0.1filter.rds")
sim_counts_mtg <- getSimulatedDataset(ps_mtg_filt, model_mtg, n = "sample_counts", divisor = 1000)
saveRDS(sim_counts_mtg, "sim_counts_mtg_2topics.rds")

model_mtt <- readRDS("../results/latent_variable_modeling/mtt_models/model_mtt_3topics_3000iter_0.1filter.rds")
sim_counts_mtt <- getSimulatedDataset(ps_mtt_filt, model_mtt, n = "sample_counts", divisor = 1000)
saveRDS(sim_counts_mtt, "../results/latent_variable_modeling/sim_counts_mtt_3topics.rds")

model_mbx <- readRDS("../results/latent_variable_modeling/mbx_models/model_mbx_10topics_10000iter_0.1filter.rds")
sim_counts_mbx <- getSimulatedDataset(ps_mbx_filt, model_mbx, n = "sample_counts")
saveRDS(sim_counts_mbx, "../results/latent_variable_modeling/sim_counts_mbx_10topics.rds")

```


# Plot Model Fit
```{r}
plotModelFit <- function(ps, sim_counts, suffix){
  tmp <- getModelFit(ps, sim_counts)
  pdf(paste0("../../results/latent_variable_modeling/model_selection/procrustes1_", suffix, ".pdf"),  width = 5, height = 5)
  p <- plot(tmp[[1]], kind = 1)
  dev.off()
  
  #pdf(paste0("../results/latent_variable_modeling/model_selection/procrustes2_", suffix, ".pdf"),  width = 7, height = 3)
  #plot(tmp[[1]], kind = 2, title = suffix)
  #dev.off()
  ggsave(paste0("../../results/latent_variable_modeling/model_selection/quantileTest_", suffix, ".pdf") ,tmp[[2]], width = 5, height = 2.3)

}
################### 16s ##############################
sim_counts_16s <- readRDS("sim_counts_16s_3topics.rds")
plotModelFit(ps_16s_filt, sim_counts_16s, "16s_3topics")

sim_counts_16s <- readRDS("sim_counts_16s_10topics.rds")
plotModelFit(ps_16s_filt, sim_counts_16s, "16s_10topics")

#3
################### MTG ##############################
sim_counts_mtg <- readRDS("../results/latent_variable_modeling/sim_counts_mtg_2topics.rds")
plotModelFit(ps_mtg_filt, sim_counts_mtg, "mtg_2topics")

sim_counts_mtg <- readRDS("sim_counts_mtg_10topics.rds")
plotModelFit(ps_mtg_filt, sim_counts_mtg, "mtg_10topics")


################### MTT ##############################
sim_counts_mtt <- readRDS("sim_counts_mtt_3topics.rds")
plotModelFit(ps_mtt_filt, sim_counts_mtt, "mtt_3topics")

sim_counts_mtt <- readRDS("sim_counts_mtt_10topics.rds")
plotModelFit(ps_mtt_filt, sim_counts_mtt, "mtt_10topics")


################### MBX ##############################
sim_counts_mbx <- readRDS("../results/latent_variable_modeling/sim_counts_mbx_3topics.rds")
plotModelFit(ps_mbx_filt, sim_counts_mbx, "mbx_3topics")

sim_counts_mbx <- readRDS("../results/latent_variable_modeling/sim_counts_mbx_4topics.rds")
plotModelFit(ps_mbx_filt, sim_counts_mbx, "mbx_4topics")

sim_counts_mbx <- readRDS("../results/latent_variable_modeling/sim_counts_mbx_6topics.rds")
plotModelFit(ps_mbx_filt, sim_counts_mbx, "mbx_6topics")

sim_counts_mbx <- readRDS("../results/latent_variable_modeling/sim_counts_mbx_10topics.rds")
plotModelFit(ps_mbx_filt, sim_counts_mbx, "mbx_10topics")

sim_counts_mbx <- readRDS("../results/latent_variable_modeling/sim_counts_mbx_20topics.rds")
plotModelFit(ps_mbx_filt, sim_counts_mbx, "mbx_20topics")

sim_counts_mbx <- readRDS("../results/latent_variable_modeling/sim_counts_mbx_28topics.rds")
plotModelFit(ps_mbx_filt, sim_counts_mbx, "mbx_28topics")

plot(c(3,4,6,10, 20, 28), c(.78, .816, .862, 0.899, .921,.93), main= "Correlation of quantile improvement with increasing model complexity")

```

```{r}

plotQuantileCorrelations <- function(ps, directory, title, sim = F, divisor = 1){
  files <- list.files(directory)
  files <- files[grepl("model", files)]
  r2s <- c()
  topics <- c()
  for(file in files){
    i <- sub(".*?_", "", file)
    i <- sub(".*?_", "", i)
    i <- sub("topics.*", "", i)
    #i <- sub("topics_3000iter_0.05filter.rds", "", i)
    i <- as.numeric(i)
    print(i)
    topics <- c(topics, i)
    model <- readRDS(paste0(directory, file))
    #likelihoods[[i]] <- model@loglikelihood
    if(sim){
      sim_counts <- getSimulatedDataset(ps, model, n = "sample_counts", divisor = divisor)
      saveRDS(sim_counts, paste0(directory, "/sim_counts_", i, "topics.rds"))
    }else{
      sim_counts <- readRDS(paste0(directory, "/sim_counts_", i, "topics.rds"))
    }
    tmp <- quantileTest(t(ps@otu_table), sim_counts)
    line <- tmp$line
    r2s <- c(r2s, summary(line)[[9]])
    rm(model)
    rm(sim_counts)
  }
  df <- data.frame(numTopics = topics, r2s = r2s)
  ggplot(df, aes(x = numTopics, y = r2s)) +
    geom_point()+ 
    geom_line()+
    theme_bw()+
    theme(axis.text = element_text(size = 14), axis.title = element_text(size = 14))+
    xlab("Number of Topics") + ylab("Quantile Correlation")+
    ggtitle(title)
}
p_16s <- plotQuantileCorrelations(ps_16s_filt, "../../results/latent_variable_modeling/16s_models_50000/", title = "16s", sim = F)
ggsave("../../results/latent_variable_modeling/16s_models_50000/quantile_correlation_num_topics.pdf", p_16s,, width = 6, height = 3.5)

p_mtg <- plotQuantileCorrelations(ps_mtg_filt, "../../results/latent_variable_modeling/mtg_models_50000/", title = "MTG", sim = T, divisor = 1000)
ggsave("../../results/latent_variable_modeling/mtg_models_50000/quantile_correlation_num_topics.pdf", p_mtg, , width = 6, height = 3.5)

p_mtt <- plotQuantileCorrelations(ps_mtt_filt, "../../results/latent_variable_modeling/mtt_models_50000/", title = "MTT", sim = T)
ggsave("../../results/latent_variable_modeling/mtt_models_50000/quantile_correlation_num_topics.pdf", p_mtt)

p_mbx <- plotQuantileCorrelations(ps_mbx_filt, "../../results/latent_variable_modeling/mbx_models_50000/", title = "MBX", sim = T)
ggsave("../../results/latent_variable_modeling/mbx_models_50000/quantile_correlation_num_topics.pdf", p_mbx, width = 6, height = 3.5)

```


# Load final models
```{r}
model_16s <- readRDS("../../results/latent_variable_modeling/16s_models_50000/model_16s_4topics_50000iter_0.05filter.rds")
model_mtg <- readRDS("../../results/latent_variable_modeling//mtg_models_50000/model_mtg_2topics_50000iter_0.1filter.rds") # 3 topics based on hist(colSums(thetas_mtg), breaks = 50) at high number of topics. Only 3 have large values across samples

model_mtt <- readRDS("../../results/latent_variable_modeling//mtt_models_50000/model_mtt_2topics_50000iter_0.1filter.rds") # 3topics, same logic
model_mbx <- readRDS("../../results/latent_variable_modeling/mbx_models_50000//model_mbx_4topics_50000iter_0.1filter.rds")


```


# Line up topics based on shared distributions across samples
```{r}
thetas_16s <- model_16s@gamma
rownames(thetas_16s) <- sample_names(ps_16s_filt)
betas_16s <- model_16s@beta
tax <- data.frame(ps_16s_filt@tax_table)
colnames(betas_16s) <- paste(tax[, 2], tax[, 7], tax[,8])
#thetas_16s <- apply(thetas_16s, 2, function(x) return((x - mean(x)) / sd(x)))
  
thetas_mtg <- model_mtg@gamma
rownames(thetas_mtg) <- sample_names(ps_mtg_filt)
betas_mtg <- model_mtg@beta
tax <- data.frame(ps_mtg_filt@tax_table)
colnames(betas_mtg) <- tax[,2]
#thetas_mtg <- apply(thetas_mtg, 2, function(x) return((x - mean(x)) / sd(x)))

thetas_mtt <- model_mtt@gamma
rownames(thetas_mtt) <- sample_names(ps_mtt_filt)
betas_mtt <- model_mtt@beta
tax <- data.frame(ps_mtt_filt@tax_table)
#colnames(betas_mtt) <- tax[,2]
#thetas_mtt <- apply(thetas_mtt, 2, function(x) return((x - mean(x)) / sd(x)))

thetas_mbx <- model_mbx@gamma
rownames(thetas_mbx) <- sample_names(ps_mbx_filt)
betas_mbx <- model_mbx@beta
tax <- data.frame(ps_mbx_filt@tax_table)
colnames(betas_mbx) <- tax[ , 2]
#thetas_mbx <- apply(thetas_mbx, 2, function(x) return((x - mean(x)) / sd(x)))


colnames(thetas_16s) <- paste0("topic", seq(1:ncol(thetas_16s)), "_16s")
colnames(thetas_mtg) <- paste0("topic", seq(1:ncol(thetas_mtg)), "_MTG")
colnames(thetas_mtt) <- paste0("topic", seq(1:ncol(thetas_mtt)), "_MTT")
colnames(thetas_mbx) <- paste0("topic", seq(1:ncol(thetas_mbx)), "_MBX")

rownames(betas_16s) <- paste0("topic", seq(1:ncol(thetas_16s)), "_16s")
rownames(betas_mtg) <- paste0("topic", seq(1:ncol(thetas_mtg)), "_MTG")
rownames(betas_mtt) <- paste0("topic", seq(1:ncol(thetas_mtt)), "_MTT")
rownames(betas_mbx) <- paste0("topic", seq(1:ncol(thetas_mbx)), "_MBX")

samples <- intersect(intersect(intersect(rownames(thetas_16s), rownames(thetas_mtg)), rownames(thetas_mtt)), rownames(thetas_mbx))

df <- data.frame(cbind(thetas_16s[samples, ], thetas_mtg[samples, ], thetas_mtt[samples, ], thetas_mbx[samples, ]))
p <- pheatmap(df, clustering_distance_cols = "correlation", clustering_distance_rows = "correlation")

clustering <- cutree(p$tree_col, k = 4)
omic <- sapply(names(clustering), function(x) return(strsplit(x, "_")[[1]][2]))
new_topic_name <- paste0("topic", as.numeric(clustering))
new_names <- paste0(new_topic_name, "_", omic)
names(new_names) <- names(clustering)

names_16s <- grepl("16s", names(new_names))
thetas_16s <- thetas_16s[ , names(new_names)[names_16s]]
colnames(thetas_16s) <- new_names[names_16s]
betas_16s <- betas_16s[names(new_names)[names_16s], ]
rownames(betas_16s) <- new_names[names_16s]

names_mtg <- grepl("MTG", names(new_names))
thetas_mtg <- thetas_mtg[ , names(new_names)[names_mtg]]
colnames(thetas_mtg) <- new_names[names_mtg]
betas_mtg <- betas_mtg[names(new_names)[names_mtg], ]
rownames(betas_mtg) <- new_names[names_mtg]

names_mtt <- grepl("MTT", names(new_names))
thetas_mtt <- thetas_mtt[ , names(new_names)[names_mtt]]
colnames(thetas_mtt) <- new_names[names_mtt]
betas_mtt <- betas_mtt[names(new_names)[names_mtt], ]
rownames(betas_mtt) <- new_names[names_mtt]

names_mbx <- grepl("MBX", names(new_names))
thetas_mbx <- thetas_mbx[ , names(new_names)[names_mbx]]
colnames(thetas_mbx) <- new_names[names_mbx]
betas_mbx <- betas_mbx[names(new_names)[names_mbx], ]
rownames(betas_mbx) <- new_names[names_mbx]


df <- data.frame(cbind(thetas_16s[samples, ], thetas_mtg[samples, ], thetas_mtt[samples, ], thetas_mbx[samples, ]))
p <- pheatmap(df, clustering_distance_cols = "correlation", clustering_distance_rows = "correlation",  color = diverging_hcl(200, palette = "PurpleGreen",rev = T))
```


```{r}
topic1 <- df[ , grepl("topic1", colnames(df))]
cor1 <- apply(topic1, 2, function(x){
  apply(topic1, 2, function(y){
    return(cor.test(x, y)$p.value)
  })
})

topic2 <- df[ , grepl("topic2", colnames(df))]
cor2 <- apply(topic2, 2, function(x){
  apply(topic2, 2, function(y){
    return(cor.test(x, y)$p.value)
  })
})

topic3 <- df[ , grepl("topic3", colnames(df))]
cor3 <- apply(topic3, 2, function(x){
  apply(topic3, 2, function(y){
    return(cor.test(x, y)$p.value)
  })
})

topic4 <- df[ , grepl("topic4", colnames(df))]
cor4 <- apply(topic4, 2, function(x){
  apply(topic4, 2, function(y){
    return(cor.test(x, y)$p.value)
  })
})

apply(cor1, 1, function(x) return(x < .05))
apply(cor2, 1, function(x) return(x < .05)) 
apply(cor3, 1, function(x) return(x < .05)) # drop topic2_MBC
apply(cor4, 1, function(x) return(x < .05)) # drop topic4_MTG, and topic4_MTT

betas_mtg <- betas_mtg[!colnames(thetas_mtg) %in% c("topic4_MTG"), ]
betas_mtt <- betas_mtt[!colnames(thetas_mtt) %in% c("topic4_MTT"), ]

thetas_mtg <- thetas_mtg[ , !colnames(thetas_mtg) %in% c("topic4_MTG")]
thetas_mtt <- thetas_mtt[ , !colnames(thetas_mtt) %in% c("topic4_MTT")]

#thetas_mbx <- thetas_mbx[ , c(2,3,4)]
df <- data.frame(cbind(thetas_16s[samples, ], thetas_mtg[samples, ], thetas_mtt[samples, ], thetas_mbx[samples, ]))

topic1 <- df[ , grepl("topic1", colnames(df))]
cor1 <- apply(topic1, 2, function(x){
  apply(topic1, 2, function(y){
    return(cor.test(x, y)$p.value)
  })
})

topic2 <- df[ , grepl("topic2", colnames(df))]
cor2 <- apply(topic2, 2, function(x){
  apply(topic2, 2, function(y){
    return(cor.test(x, y)$p.value)
  })
})

topic3 <- df[ , grepl("topic3", colnames(df))]
cor3 <- apply(topic3, 2, function(x){
  apply(topic3, 2, function(y){
    return(cor.test(x, y)$p.value)
  })
})

topic4 <- df[ , grepl("topic4", colnames(df))]
cor4 <- apply(topic4, 2, function(x){
  apply(topic4, 2, function(y){
    return(cor.test(x, y)$p.value)
  })
})

apply(cor1, 1, function(x) return(x < .05))
apply(cor2, 1, function(x) return(x < .05)) 
apply(cor3, 1, function(x) return(x < .05)) 
apply(cor4, 1, function(x) return(x < .05))
```



# Correlate Metadata
```{r}
palette = "Purple-Green"
width = 4

thetas_16s <- thetas_16s[, !is.na(colSums(thetas_16s))]
res_16s <- getCorrelationsMetadata(ps_16s_filt, thetas_16s, ending = "16s", n = ncol(thetas_16s))
rvals_16s <- res_16s$rvals
pvals_16s <- res_16s$p_adj
p <- pheatmap(t(rvals_16s), cluster_rows = F, color = diverging_hcl(n = 300, palette = palette, rev = T, alpha = 0.8))
#ggsave("../../results/latent_variable_modeling/metadata_corr_16s.pdf", p, width = width, height = 6)

thetas_mtg <- thetas_mtg[, !is.na(colSums(thetas_mtg))]
res_mtg <- getCorrelationsMetadata(ps_mtg_filt, thetas_mtg, ending = "MTG",n = ncol(thetas_mtg))
rvals_mtg <- res_mtg$rvals
pvals_mtg <- res_mtg$p_adj
p <- pheatmap(t(rvals_mtg), cluster_rows = F, color = diverging_hcl(n = 300, palette = palette, rev = T, alpha = 0.8))
#ggsave("../../results/latent_variable_modeling/metadata_corr_mtg.pdf", p, width = width, height = 6)

thetas_mtt <- thetas_mtt[, !is.na(colSums(thetas_mtt))]
res_mtt <- getCorrelationsMetadata(ps_mtt_filt, thetas_mtt, ending = "MTT", n = ncol(thetas_mtt))
rvals_mtt <- res_mtt$rvals
pvals_mtt <- res_mtt$p_adj
p <- pheatmap(t(rvals_mtt[!is.na(rowSums(rvals_mtt)), ]), cluster_rows = F, color = diverging_hcl(n = 300, palette = palette, rev = T, alpha = 0.8))
#ggsave("../../results/latent_variable_modeling/metadata_corr_mtt.pdf", p, width = width, height = 6)

thetas_mbx <- thetas_mbx[, !is.na(colSums(thetas_mbx))]
res_mbx <- getCorrelationsMetadata(ps_mbx_filt, thetas_mbx, ending = "MBX", n = ncol(thetas_mbx))
rvals_mbx <- res_mbx$rvals
pvals_mbx <- res_mbx$p_adj
p <- pheatmap(t(rvals_mbx[!is.na(rowSums(rvals_mbx)), ]), cluster_rows = F, color = diverging_hcl(n = 300, palette = palette, rev = T, alpha = 0.8))
#ggsave("../../results/latent_variable_modeling/metadata_corr_mbx.pdf", p, width = width, height = 6)

df <- data.frame(rbind(rvals_16s, rvals_mtg, rvals_mtt[!is.na(rowSums(rvals_mtt)), ], rvals_mbx))
df_p <- data.frame(rbind(pvals_16s, pvals_mtg, pvals_mtt[!is.na(rowSums(pvals_mtt)), ], pvals_mbx))
df_sig <- df_p < .1
df_annotate <- apply(df_sig, 1, function(x) ifelse(x, "*", ""))

p <- pheatmap(t(df), cluster_rows = T, cluster_cols = F,
              clustering_distance_rows = "correlation",
              color = diverging_hcl(n = 300, palette = palette, rev = T, alpha = 0.8),
              display_numbers = df_annotate)
#ggsave("../../results/latent_variable_modeling/metadata_corr_all.pdf", p, width = 6, height =5)
```



# Replot fit MBX model
```{r}
setClass("model", slots = list(gamma = "matrix", beta = "matrix"))
model_tmp <- new("model", gamma = as.matrix(thetas_mbx), beta = as.matrix(betas_mbx))
sim_counts_mbx <- getSimulatedDataset(ps_mbx_filt, model_tmp, n = "sample_counts")
plotModelFit(ps_mbx_filt, sim_counts_mbx, "mbx_3topics")
```




# Full heatmap
```{r, fig.height=10}
samples <- intersect(intersect(intersect(rownames(thetas_16s), rownames(thetas_mtg)), rownames(thetas_mtt)), rownames(thetas_mbx))

df <- data.frame(cbind(thetas_16s[samples, ], thetas_mtg[samples, ], thetas_mtt[samples, ], thetas_mbx[samples, ]))
df <- df[, !is.na(colSums(df))]
phen <- data.frame('phenotype' = ps_16s_filt@sam_data[rownames(df), 'phenotype'])
p <- pheatmap(t(df), annotation_col = phen, cluster_rows = T, 
              clustering_distance_rows = 'correlation',clustering_distance_cols = 'correlation',
              color = diverging_hcl(200, palette = "PurpleGreen",rev = T))

clustering <- cutree(p$tree_col, k = 2)
names(clustering) == rownames(df)



df_phen <- df
df_phen$phenotype <- phen$phenotype
df_phen$phenotype <- ifelse(df_phen$phenotype == "A", 1, 0)

data1 <- df_phen[clustering == 1, ]
data2 <- df_phen[clustering == 2, ]
#data3 <- df_phen[clustering == 3, ]
#data4 <- df_phen[clustering == 4, ]

dim(data1)
dim(data2)
#dim(data3)
#dim(data4)

#ggsave("../../results/latent_variable_modeling/full_heatmap.pdf", p, width = 14, height = 7)
```


# Differences in metadata
```{r}
control_variables <- c("whole_grain", "olive_oil", "fat_oil_freq",  "dairy", "dairy_freq", "fruit", "fruit_freq", "vegetable", "vegetable_freq", "fermented_vegetables", "meal_home_prep", "meal_ready_eat",  "restaurant", "meat", "seafood",  "sweetened_drink",  "sugary_food", "starchy_food",    "csection")

plotMetadata <- function(data, control_variables){
  meta <- as.data.frame(ps_mtg@sam_data[rownames(data), control_variables])
  tmp <- apply(meta, 2, function(x){
    if("True" %in% x){
      x <- as.logical(x)
    }
    x <- as.numeric(x)
    return(x)
  } )
  meta <- data.frame(tmp)
  meta$phenotype <- as.factor(ifelse(data$phenotype == 1, "A", "N"))
  
  p <- ggplot(melt(meta, id.vars = c("phenotype")), aes(x = phenotype, y = value, fill = phenotype)) +
    geom_boxplot() +
    geom_jitter(width = 0.1)+
    stat_compare_means(aes(label=..p.adj..), method = "wilcox.test", comparisons = list(c("A", "N")))+
    theme_bw()+
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
    ggtitle("Within cluster 1") + facet_wrap(~variable, scales = 'free') +
    scale_y_continuous(expand = expansion(mult = c(0, 0.2)))+
    scale_fill_manual(values = c(asd_color, td_color))
  return(p)


}
p <- plotMetadata(data1, control_variables)
ggsave("../../results/latent_variable_modeling/metadata_diff_cluster1.pdf", p, width = 11, height = 10)
p <- plotMetadata(data2, control_variables)
ggsave("../../results/latent_variable_modeling/metadata_diff_cluster2.pdf", p, width = 11, height = 10)
#p <- plotMetadata(data3, control_variables)
#ggsave("../../results/latent_variable_modeling/metadata_diff_cluster3.pdf", p, width = 11, height = 10)
df_meta <- data.frame(ps_mtg@sam_data)
df_meta$phenotype <- ifelse(df_meta$phenotype == "A", 1, 0)
p <- plotMetadata(df_meta, control_variables)
ggsave("../../results/latent_variable_modeling/metadata_diff_cluster_none.pdf", p, width = 11, height = 10)


```



# Cluster samples w/ phenotype and topics
```{r}
df <- data.frame(cbind(thetas_16s[samples, ], thetas_mtg[samples, ], thetas_mtt[samples, ], thetas_mbx[samples, ]))
df <- df[, !is.na(colSums(df))]

df_phen <- df
df_phen$phenotype <- phen$phenotype
df_phen$phenotype <- ifelse(df_phen$phenotype == "A", 1, 0)
df_plot <- df_phen
rownames(df_plot) <- as.character(unlist(ps_mtg@sam_data[rownames(df), "host_name"]))
rownames(phen) <- as.character(unlist(ps_mtg@sam_data[rownames(df), "host_name"]))

quantile_breaks <- function(xs, n = 10) {
    breaks <- quantile(xs, probs = seq(0, 1, length.out = n))
    return(breaks[!duplicated(breaks)])
}

mat_breaks <- quantile_breaks(unlist(df_plot), n = 100)

p <- pheatmap(t(df_phen), cluster_rows = T,
              color = diverging_hcl(100, palette = "PurpleGreen",rev = T, alpha = 0.8),
              breaks = mat_breaks,
              fontsize_row = 16)

ggsave("../../results/latent_variable_modeling/subtyping.pdf", p, height = 5, width = 10)
clustering <- cutree(p$tree_col, k = 2)
names(clustering) == rownames(df)
data1 <- df_phen[clustering == 1, ]
data2 <- df_phen[clustering == 2, ]
dim(data1)
dim(data2)
#data3 <- df_phen[clustering == 3, ]
```

# Set clusters
```{r}

df_phen$cluster <- rep(NA, nrow(df_phen)) 
df_phen$cluster[rownames(df_phen) %in% rownames(data1)]= "cluster1"
df_phen$cluster[rownames(df_phen) %in% rownames(data2)]= "cluster2"
#df_phen$cluster[rownames(df_phen) %in% rownames(data3)]= "cluster3"
df_phen$cluster <- as.factor(df_phen$cluster)
df_phen <- df_phen[!is.na(df_phen$cluster), ]
```




# Colors
```{r}
asd_color <- "#01898D"
td_color <- "#67A102"

col1 <- "#DB5461"
col2 <- "#AE8E1C"
col3 <- "blue"
```

# Topic differences between groups
```{r}
out <- Boruta(df_phen %>% select(-c("cluster", "phenotype")), df_phen$cluster)
out
df_plot <- df_phen
df_plot$cluster <- df_phen$cluster

pvals <- apply(df_plot  %>% select(-c("phenotype", "cluster")), 2, function(x){
  return(wilcox.test(x[df_plot$cluster == "cluster1"], x[df_plot$cluster == "cluster2"])$p.value)
})

df_plot <- df_plot[ , names(pvals)[pvals < .05]]
df_plot$cluster <- df_phen$cluster
df_plot_melt <- melt(df_plot, id.vars = c("cluster"))
df_plot_melt$variable <- factor(df_plot_melt$variable, levels = c("topic1_16s", "topic2_16s", "topic3_16s", "topic4_16s",
                                                                  "topic1_MTG", "topic2_MTG", "topic3_MTG", "topic4_MTG",
                                                                  "topic1_MTT",  "topic2_MTT",  "topic3_MTT",  "topic4_MTT",
                                                                  "topic1_MBX", "topic2_MBX","topic3_MBX", "topic4_MBX" ))
p <- ggplot(df_plot_melt, aes(x = variable, y = value, fill = cluster)) +
  geom_boxplot() + stat_compare_means(aes(label=..p.format..), method = "wilcox.test")+
  xlab("")+
  theme_bw() +theme(axis.text.x = element_text(angle = 70, vjust = 0.9, hjust=1, size = 12),
                    legend.text = element_text(size = 11)) +
  scale_fill_manual(values = c(col1, col2, col3), name = "")+
  ggtitle("To Cluster")+
  scale_y_continuous(expand = expansion(mult = c(0, 0.09)))
#ggsave("../../results/latent_variable_modeling/subsetting_topics.pdf", p, width = 9.6, height = 5)
```

```{r}
pvalue_filter <- function(df_plot){
  pvals <- apply(df_plot %>% select(-c("phenotype")), 2, function(x){
    aut <- df_plot$phenotype == "A"
    nt <- df_plot$phenotype == "N"
    return(wilcox.test(x[aut], x[nt])$p.value)
  } )
  pvals <- p.adjust(pvals, method = "fdr")
  keep <- colnames(df_plot %>% select(-c("phenotype")))[pvals<.05]
  keep <- c(keep, "phenotype")
  return(list(df_plot = df_plot[, keep], pvals = pvals))
}

doThePlot <- function(df_plot, title, ncol = 4){
  p1 <- ggplot(melt(df_plot, id.vars = c("phenotype")), aes(x = phenotype, y = value, fill = phenotype)) +
  geom_boxplot(outlier.alpha = 0.5, outlier.color = "grey") +
  geom_jitter(width = 0.01) + 
  stat_compare_means(aes(label=paste("p = ", ..p.adj..)), method = "wilcox.test", size = 3.5)+
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  facet_wrap(~variable, scales = 'free', ncol = ncol, labeller = label_wrap_gen()) +
  scale_y_continuous(expand = expansion(mult = c(0, 0.2))) +
  theme_bw()+
  scale_fill_manual(values = c(asd_color, td_color))+
  ggtitle(title)
}

```


# MBX difference between groups
```{r}

################### Group 1  ####################################################################
res1_mbx <- getImportantFactors(ps_mbx, data1, iter = iter)
#saveRDS(res1_mbx, "../../results/latent_variable_modeling/phenotype_diff_res_mbx_group1.rds")
#res1_mbx <- readRDS("../../results/latent_variable_modeling/phenotype_diff_res_mbx_group1.rds")
table(res1_mbx$imp)
#thresh <- quantile(table(res1_mbx$imp), probs = seq(0, 1, 0.01))["80%"]
thresh <- 5
df_plot <- keepMostImp(res1_mbx$df_plot, res1_mbx$imp, thresh)
tmp <- pvalue_filter(df_plot)
df_plot <- tmp$df_plot
pvals<- tmp$pvals
p1 <- doThePlot(df_plot, title = "Within cluster 1", ncol = 3)
#ggsave("../../results/latent_variable_modeling/mbx_cluster1_diff.pdf", p1, width = 8, height= 1.833333*1 + .5)
p1

################### Group 2  ####################################################################
res2_mbx <- getImportantFactors(ps_mbx, data2, iter = iter)
#saveRDS(res2_mbx, "../../results/latent_variable_modeling/phenotype_diff_res_mbx_group2.rds")
#res2_mbx <- readRDS("../../results/latent_variable_modeling/phenotype_diff_res_mbx_group2.rds")
table(res2_mbx$imp)
#thresh <- quantile(table(res2_mbx$imp), probs = seq(0, 1, 0.01))["80%"]
thresh <- 3
df_plot <- keepMostImp(res2_mbx$df_plot, res2_mbx$imp, thresh)
tmp <- pvalue_filter(df_plot)
df_plot <- tmp$df_plot
pvals<- tmp$pvals

colnames(df_plot) <- sub("propyl\\.", "propyl\\.\n", colnames(df_plot))
colnames(df_plot) <- sub("straight\\.", "\n straight", colnames(df_plot))
p2 <- doThePlot(df_plot, title = "Within cluster 2", ncol = 3)
#ggsave("../../results/latent_variable_modeling/mbx_cluster2_diff.pdf", p2, width = 2.5 * 2 + 0.5, height= 1.833333*1 + .5)
 p2


```

# MTG difference between groups
```{r}
#iter = 100
#thresh = 20
################### Group 1  ####################################################################
res1_mtg <- getImportantFactors(ps_mtg, data1, iter = iter)
saveRDS(res1_mtg, "../../results/latent_variable_modeling/phenotype_diff_res_mtg_group1.rds")
#res1_mtg <- readRDS("../../results/latent_variable_modeling/phenotype_diff_res_mtg_group1.rds")
table(res1_mtg$imp)
thresh <- quantile(table(res1_mtg$imp), probs = seq(0, 1, 0.01))["95%"]
df_plot <- keepMostImp(res1_mtg$df_plot, res1_mtg$imp, thresh)
tmp <- pvalue_filter(df_plot)
df_plot <- tmp$df_plot
pvals<- tmp$pvals

old_names <- colnames(df_plot %>% select(-c("phenotype")))
new_names <- as.character(data.frame(ps_mtg_filt@tax_table)[old_names , 2])
colnames(df_plot) <- c(paste(new_names , old_names),"phenotype")
#colnames(df_plot) <- sub(",", "\n", colnames(df_plot))
#colnames(df_plot) <- sub("system ATP", "system\nATP",colnames(df_plot))
#colnames(df_plot) <- sub("\\[", "\n\\[", colnames(df_plot))
#colnames(df_plot) <- sub("transporting", "transporting\n",colnames(df_plot))


p1 <- doThePlot(df_plot, title = "Within cluster 1", ncol = 3)
ggsave("../../results/latent_variable_modeling/mtg_cluster1_diff.pdf", p1, width = 9, height= 1.833333*2 + .5)
p1

################### Group 2  ####################################################################
res2_mtg <- getImportantFactors(ps_mtg, data2, iter = iter)
saveRDS(res2_mtg, "../../results/latent_variable_modeling/phenotype_diff_res_mtg_group2.rds")
#res2_mtg <- readRDS("../../results/latent_variable_modeling/phenotype_diff_res_mtg_group2.rds")
#thresh <- 15
table(res2_mtg$imp)
#thresh <- quantile(table(res2_mtg$imp), probs = seq(0, 1, 0.01))["95%"]
thresh <- 5
df_plot <- keepMostImp(res2_mtg$df_plot, res2_mtg$imp, thresh)
tmp <- pvalue_filter(df_plot)
df_plot <- tmp$df_plot
pvals<- tmp$pvals
 
old_names <- colnames(df_plot %>% select(-c("phenotype")))
new_names <- as.character(ps_mtg_filt@tax_table[ old_names , 2])
colnames(df_plot) <- c(paste(new_names , old_names),"phenotype")


#df_plot <- df_plot[ , ! colnames(df_plot) %in% c("peptidoglycan pentaglycine glycine transferase (the second and third glycine) \n[EC:2.3.2.17] K11694", "UDP-2-acetamido-2-deoxy-ribo-hexuluronate aminotransferase \n[EC:2.6.1.98] K13017")]

p2 <- doThePlot(df_plot, title = "Within cluster 2", ncol = 3)
ggsave("../../results/latent_variable_modeling/mtg_cluster2_diff.pdf", p2, width = 8, height= 1.833333*2 + .5)
p2


```

# MTT difference between groups
```{r}

################### Group 1  ####################################################################
res1_mtt <- getImportantFactors(ps_mtt, data1, iter = iter)
saveRDS(res1_mtt, "../../results/latent_variable_modeling/phenotype_diff_res_mtt_group1.rds")
res1_mtt <- readRDS("../../results/latent_variable_modeling/phenotype_diff_res_mtt_group1.rds")
table(res1_mtt$imp)
thresh <- quantile(table(res1_mtt$imp), probs = seq(0, 1, 0.01))["95%"]
df_plot <- keepMostImp(res1_mtt$df_plot, res1_mtt$imp, thresh)

tmp <- pvalue_filter(df_plot)
df_plot <- tmp$df_plot
pvals<- tmp$pvals

old_names <- colnames(df_plot %>% select(-c("phenotype")))
new_names <- as.character(ps_mtt_filt@tax_table[ old_names , 2])
colnames(df_plot) <- c(paste(new_names , old_names),"phenotype")
colnames(df_plot) <- sub(",", "\n", colnames(df_plot))
colnames(df_plot) <- sub("\\[", "\n\\[", colnames(df_plot))
colnames(df_plot) <- sub("reductase", "reductase\n", colnames(df_plot))
colnames(df_plot) <- sub("outer", "outer\n", colnames(df_plot))

p1 <- doThePlot(df_plot, "Within Cluster 1", ncol = 3)
ggsave("../../results/latent_variable_modeling/mtt_cluster1_diff.pdf", p1, width = 8, height= 1.833333*1 + .5)
p1

################### Group 2  ####################################################################

res2_mtt <- getImportantFactors(ps_mtt, data2, iter = iter)
saveRDS(res2_mtt, "../../results/latent_variable_modeling/phenotype_diff_res_mtt_group2.rds")
res2_mtt <- readRDS("../../results/latent_variable_modeling/phenotype_diff_res_mtt_group2.rds")
table(res2_mtt$imp)
thresh <- quantile(table(res2_mtt$imp), probs = seq(0, 1, 0.01))["95%"]
df_plot <- keepMostImp(res2_mtt$df_plot, res2_mtt$imp, thresh)
tmp <- pvalue_filter(df_plot)
df_plot <- tmp$df_plot
pvals<- tmp$pvals

old_names <- colnames(df_plot %>% select(-c("phenotype")))
new_names <- as.character(ps_mtt_filt@tax_table[ old_names , 2])
colnames(df_plot) <- c(paste(new_names , old_names),"phenotype")


p2 <- doThePlot(df_plot, title = "Within Cluster 2", ncol = 3)
ggsave("../../results/latent_variable_modeling/mtt_cluster2_diff.pdf", p2, width = 8, height= 1.833333*1 + .5)
p2
```


# 16s: differ between groups 
```{r}

res1_16s <- getImportantFactors(ps_16s, data1, iter = iter)
saveRDS(res1_16s, "../../results/latent_variable_modeling/phenotype_diff_res_16s_group1.rds")
res1_16s <- readRDS("../../results/latent_variable_modeling/phenotype_diff_res_16s_group1.rds")
table(res1_16s$imp)
thresh <- quantile(table(res1_16s$imp), probs = seq(0, 1, 0.01))["90%"]
df_plot <- keepMostImp(res1_16s$df_plot, res1_16s$imp, thresh)
tmp <- pvalue_filter(df_plot)
df_plot <- tmp$df_plot
pvals<- tmp$pvals

old_names <- colnames(df_plot %>% select(-c("phenotype")))
new_names <- paste(as.character(ps_16s@tax_table[old_names, 2]),
                         as.character(ps_16s@tax_table[old_names, 6]),
                         as.character(ps_16s@tax_table[old_names, 8]))
colnames(df_plot) <- c(new_names,"phenotype")
colnames(df_plot) <- sub(" ", "\n", colnames(df_plot))
df_plot <- df_plot[, ! (colnames(df_plot) %in% c("Bacteroidetes\nBacteroides caccae", "Firmicutes\nNA NA.1"))]
colnames(df_plot) <- gsub("_group NA", "", colnames(df_plot))

p1 <- doThePlot(df_plot, title= "Within Cluster 1", ncol = 3)
ggsave("../../results/latent_variable_modeling/16s_cluster1_diff.pdf", p1, width = 8, height = 1.833333*1 + .5)




res2_16s <- getImportantFactors(ps_16s, data2, iter = iter)
saveRDS(res2_16s, "../../results/latent_variable_modeling/phenotype_diff_res_16s_group2.rds")
res2_16s <- readRDS("../../results/latent_variable_modeling/phenotype_diff_res_16s_group2.rds")
table(res2_16s$imp)
thresh <- quantile(table(res1_16s$imp), probs = seq(0, 1, 0.01))["90%"]
df_plot <- keepMostImp(res2_16s$df_plot, res2_16s$imp, thresh)
tmp <- pvalue_filter(df_plot)
df_plot <- tmp$df_plot
pvals<- tmp$pvals

old_names <- colnames(df_plot %>% select(-c("phenotype")))
new_names <- paste(as.character(ps_16s@tax_table[old_names, 2]),
                         as.character(ps_16s@tax_table[old_names, 6]),
                         as.character(ps_16s@tax_table[old_names, 8]))
colnames(df_plot) <- c(new_names,"phenotype")
colnames(df_plot) <- sub(" ", "\n", colnames(df_plot))
#df_plot <- df_plot[, ! (colnames(df_plot) %in% c("Bacteroidetes\nBacteroides uniformis", "Firmicutes\nNA NA"   , "Firmicutes\nGranulicatella NA", "Firmicutes\nMarvinbryantia NA.1"))]
p2 <- doThePlot(df_plot, title = "Within Cluster 2", ncol = 3)
    ggsave("../../results/latent_variable_modeling/16s_cluster2_diff.pdf", p2, width = 8, height = 1.833333*1 + .5)



#res3_16s <- getImportantFactors(ps_16s, data3, iter = iter)
#saveRDS(res3_16s, "../results/latent_variable_modeling/phenotype_diff_res_16s_group3.rds")
#res3_16s <- readRDS("../results/latent_variable_modeling/phenotype_diff_res_16s_group3.rds")
#table(res3_16s$imp)
#df_plot <- keepMostImp(res3_16s$df_plot, res3_16s$imp, thresh)
#if(ncol(df_plot) > 1){
#  old_names <- colnames(df_plot %>% select(-c("phenotype")))
#  new_names <- paste(as.character(ps_16s@tax_table[old_names, 3]),
#                           as.character(ps_16s@tax_table[old_names, 6]),
#                           as.character(ps_16s@tax_table[old_names, 8]))
#  colnames(df_plot) <- c(new_names,"phenotype")
#  colnames(df_plot) <- sub(" ", "\n", colnames(df_plot))
#  p3 <- ggplot(melt(df_plot, id.vars = c("phenotype")), aes(x = phenotype, y = log(value), fill = phenotype)) +
#    geom_boxplot(outlier.alpha = 0.5, outlier.color = "grey") + geom_jitter(width = 0.03) + 
#    stat_compare_means(aes(label=paste("p = ", ..p.adj..)), method = "wilcox.test", size = 3.5)+     theme_bw() +theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
#    ggtitle("Within cluster 3") + facet_wrap(~variable, scales = 'free') +
#    scale_y_continuous(expand = expansion(mult = c(0, 0.5)))+
#    scale_fill_manual(values = c(asd_color, td_color))
    #ggsave("../results/latent_variable_modeling/16s_cluster3_diff.pdf", p3, width = 11)

#}

```





# Severity of ASD in groups
```{r}
df_phen$MARA <- ps_mtg@sam_data[rownames(df_phen), ]$MARA
ggplot(df_phen, aes(x = cluster, y = MARA, fill = cluster)) + geom_boxplot()+
  stat_compare_means()
```

# Feature weights MBX
```{r}
betas <- exp(betas_mbx)[!is.na(rowSums(betas_mbx)), ]
#diffs1 <- abs((betas[1,] - betas[2,])^2)
#diffs2 <- abs((betas[1,] - betas[3,])^2)
#diffs3 <- abs((betas[2,] - betas[3,])^2)
#df_tmp <- data.frame(diffs1, diffs2, diffs3)
#diffs <- apply(df_tmp, 1, max)

#limit <- quantile(diffs, probs = seq(0,1, .001))['98.0%']
#keep <- diffs > limit
thresh <- apply(betas, 1, function(x){
  return(quantile(x, probs = seq(0, 1, .001))["99.0%"])
})
keep <- betas[1,] > thresh[1] | betas[2,] > thresh[2] | betas[3,] > thresh[3] | betas[4, ] > thresh[4]
to_plot <- t(betas[, keep])
to_plot <- to_plot[ ,c("topic1_MBX", "topic2_MBX", "topic3_MBX", "topic4_MBX")]
quantile_breaks <- function(xs, n = 10) {
    breaks <- quantile(xs, probs = seq(0, 1, length.out = n))
    return(breaks[!duplicated(breaks)])
}

mat_breaks <- quantile_breaks(unlist(to_plot), n = 10) #lower more green

p <- pheatmap(to_plot, cluster_cols = F,
         color = diverging_hcl(50, palette = "PurpleGreen", rev = T, alpha = 0.7), #high more green
         #breaks = mat_breaks,
         fontsize_col = 14,
         fontsize_row = 14,
         fontsize = 14)
p
#p <- pheatmap(to_plot, breaks = c(seq(min(to_plot), max(to_plot[to_plot < .01]), 0.0001), max(to_plot)), cluster_cols = F,color = diverging_hcl(20, palette = "PurpleGreen",rev = T))
ggsave("../../results/latent_variable_modeling/topic_features_mbx.pdf", p, width = 8, height = 9)
```

# Feature weights MTG
```{r}
betas <- exp(betas_mtg)
betas <- betas[!is.na(rowSums(betas)), ]
thresh <- apply(betas, 1, function(x){
  return(quantile(x, probs = seq(0, 1, .001))["99.9%"])
})
keep <- betas[1,] > thresh[1] | betas[2,] > thresh[2] | betas[3,] > thresh[3]
#diffs <- abs((betas[1,] - betas[2,])^2)

#limit <- quantile(diffs, probs = seq(0,1, .001))['99.8%']
#keep <- diffs > limit
to_plot <- t(betas[, keep])
to_plot <- to_plot[, c("topic1_MTG", "topic2_MTG", "topic3_MTG")]
p <- pheatmap(to_plot, cluster_cols =F,color = diverging_hcl(20, palette = "PurpleGreen",rev = T, alpha = 0.8))

#p <- pheatmap(to_plot, breaks = c(seq(min(to_plot), max(to_plot[to_plot <= .03]), 0.00025), max(to_plot)), cluster_cols = F,color = diverging_hcl(20, palette = "PurpleGreen",rev = T))

#to_plot <- to_plot[, c("topic1_MTG", "topic3_MTG")]
#quantile_breaks <- function(xs, n = 10) {
#    breaks <- quantile(xs, probs = seq(0, 1, length.out = n))
#    return(breaks[!duplicated(breaks)])
#}

mat_breaks <- quantile_breaks(unlist(to_plot), n = 120)


p <- pheatmap(to_plot, cluster_cols = F,
              color = diverging_hcl(124, palette = "PurpleGreen", rev = T, alpha = 0.8),
              breaks = mat_breaks,
              fontsize_col = 14,
              fontsize_row = 12)


p
ggsave("../../results/latent_variable_modeling/topic_features_mtg.pdf", p, width = 7, height = 5)

```

# Feature weights MTT
```{r}
colnames(betas_mtt) <- data.frame(ps_mtt_filt@tax_table)$Definition
betas <- exp(betas_mtt)[!is.na(rowSums(betas_mtt)), ]
thresh <- apply(betas, 1, function(x){
  return(quantile(x, probs = seq(0, 1, .001))["99.9%"])
})
keep <- betas[1,] > thresh[1] | betas[2,] > thresh[2] | betas[3,] > thresh[3]

#diffs <- abs((betas[1,] - betas[2,])^2)

#limit <- quantile(diffs, probs = seq(0,1, .001))['99.8%']
#keep <- diffs > limit
to_plot <- t(betas[, keep])
p <- pheatmap(to_plot, color = diverging_hcl(20, palette = "PurpleGreen",rev = T, alpha = 0.8))

mat_breaks <- quantile_breaks(unlist(to_plot), n = 120)


p <- pheatmap(to_plot, cluster_cols = F,
              color = diverging_hcl(116, palette = "PurpleGreen", rev = T, alpha = 0.8),
              breaks = mat_breaks,
              fontsize_col = 14,
              fontsize_row = 12)



ggsave("../../results/latent_variable_modeling/topic_features_mtt.pdf", p, width = 8, height = 7)
```

# Feature weights 16s
```{r, height = 10}
betas <- exp(betas_16s)
thresh <- apply(betas, 1, function(x){
  return(quantile(x, probs = seq(0, 1, .001))["99.6%"])
})
keep <- betas[1,] > thresh[1] | betas[2,] > thresh[2] | betas[3,] > thresh[3] | betas[4,] > thresh[4]
#betas <- betas[ , keep]
#diffs1 <- abs((betas[1,] - betas[2,])^2)
#diffs2 <- abs((betas[1,] - betas[3,])^2)
#diffs3 <- abs((betas[2,] - betas[3,])^2)
#df_tmp <- data.frame(diffs1, diffs2, diffs3)
#diffs <- apply(df_tmp, 1, max)


#limit <- quantile(diffs, probs = seq(0,1, .01))['98%']
#keep <- diffs > limit
to_plot <- t(betas[, keep])
p <- pheatmap(to_plot, cluster_cols = F, 
              color = diverging_hcl(20, palette = "PurpleGreen", rev = T, alpha = 0.8),
              fontsize_col = 14,
              fontsize_row = 12)
ggsave("../../results/latent_variable_modeling/topic_features_16s.pdf", p, width = 6, height = 8)
```

# classifier
```{r}
test_accuracy <- c()
for(i in seq(1,20)){
  set.seed(i)
  df <- data.frame(cbind(thetas_16s[samples, ], thetas_mtg[samples, ], thetas_mtt[samples, ], thetas_mbx[samples, ]))
  df$phenotype <- unlist(ps_mtg@sam_data[rownames(df), "phenotype"])
  df$phenotype <- ifelse(df$phenotype == "A", 1, 0)
  df$familyID <- unlist(ps_mtg@sam_data[rownames(df), "familyID"])
  
  library(glmnet)
  library(caret)
  test_family_ids <- as.character(sample(unlist(unique(df$familyID)), 10))
  test <- df[df$familyID %in% test_family_ids, ]
  train <- df[!df$familyID %in% test_family_ids, ]
  
  res <- cv.glmnet(as.matrix(train %>% select(-c("phenotype", "familyID"))), y = train$phenotype)
  
  i = 60
  # Train data 
  probs <- predict(res, newx = as.matrix(train %>% select(-c("phenotype", "familyID"))), s = res$lambda[i])
  preds <- probs > .5
  sum(as.numeric(preds) == train$phenotype) / nrow(train)
  
  # Test data
  probs <- predict(res, newx = as.matrix(test %>% select(-c("phenotype", "familyID"))), s = res$lambda[i])
  preds <- probs > .5
  acc_test <- sum(as.numeric(preds) == test$phenotype) / nrow(test)
  test_accuracy <- c(test_accuracy, acc_test)
}

median(test_accuracy)
```


# gsea
```{r}
getRankedList <- function(ps, data){
  df <- data.frame(t(ps@otu_table[ , rownames(data) ]))
  df$phenotype <- ps@sam_data[rownames(data), ]$phenotype
  ranks <- apply(df %>% select(-c("phenotype")), 2, function(x){
    return(t.test(x[df$phenotype == "A"], x[df$phenotype == "N"])$statistic)
  })
  pvals <- apply(df %>% select(-c("phenotype")), 2, function(x){
    return(t.test(x[df$phenotype == "A"], x[df$phenotype == "N"])$p.value)
  })
  
  df <- data.frame(genes = names(ranks), imp = as.numeric(ranks))
  df <- df[order(df$imp, decreasing = T), ]
  return(df)
}


df_mtg_cluster1 <- getRankedList(ps_mtg_filt, data1)
write.csv(df_mtg_cluster1, "../../results/latent_variable_modeling/rank_list_mtg_cluster1.rnk", row.names = F, col.names = F)


df_mtt_cluster1 <- getRankedList(ps_mtt_filt, data1)
write.csv(df_mtt_cluster1, "../../results/latent_variable_modeling/rank_list_mtt_cluster1.rnk", row.names = F, col.names = F)



df_mtg_cluster2 <- getRankedList(ps_mtg_filt, data2)
write.csv(df_mtg_cluster2, "../../results/latent_variable_modeling/rank_list_mtg_cluster2.rnk", row.names = F, col.names = F)


df_mtt_cluster2 <- getRankedList(ps_mtt_filt, data2)
write.csv(df_mtt_cluster2, "../../results/latent_variable_modeling/rank_list_mtt_cluster2.rnk", row.names = F, col.names = F)
```

# Interpret gsea
```{r}
# Gsea based on a t test between ASD and TD within clusters
#Cluster 1 TD kids have secondary bile acid synthesis going on
# cluster 2 TD kids have "Ribosome" what does it mean
library(KEGGREST)
library(pathview)
mtg_res <- read.csv("../../results/gsea/GSEA-main/GSEA-main/gsea_mtg_cluster2.csv")
mtt_res <- read.csv("../../results/gsea/GSEA-main/GSEA-main/gsea_mtt_cluster2.csv")
rownames(mtg_res) <- mtg_res$Term
rownames(mtt_res) <- mtt_res$Term

mtg_res <- mtg_res[mtg_res$fdr < .05, ]
mtt_res <- mtt_res[mtt_res$fdr < .05, ]
paths <- intersect(mtg_res$Term, mtt_res$Term)

mtg_res <- mtg_res[paths,]
mtt_res <- mtt_res[paths, ]

full_names <- lapply(mtg_res$Term, function(x){ return(keggGet(x)[[1]]$NAME)})
mtg_res$full_name <- unlist(full_names)
full_names <- lapply(mtt_res$Term, function(x){ return(keggGet(x)[[1]]$NAME)})
mtt_res$full_name <- unlist(full_names)

print(mtg_res[paths, ])
print(mtt_res[paths, ])

path <- mtg_res$Term
gene_list <- mtg_res$ledge_genes
gene_list <- strsplit(gene_list, ";")[[1]]

gene_list <- mtt_res$ledge_genes
gene_list <- strsplit(gene_list, ";")[[1]]

df <- df_mtt_cluster2
rownames(df) <- df$genes
df <- df %>% select(-c(genes))

#manual colors, because pathview seems to not work with this particular pathway
df <- df[gene_list, , drop = F]
#all negative, which means all contribute to TD
df$imp <- abs(df$imp)
df$imp <- round(df$imp * 30)

color_spectrum <- colorRampPalette(c("white", td_color))
color_spectrum <- color_spectrum(max(df$imp))
cols <- color_spectrum[df$imp]
df$cols <- cols

tmp <- df %>% select(-c("imp"))
tmp

#copy/paste to https://www.genome.jp/kegg/mapper/color.html
)











```

# Silly
```{r}
control_variables <- c( "fat_oil_freq", "fruit_freq")
                       
tmp <- data.frame(ps_mtg_filt@sam_data)[samples , control_variables]
tmp <- data.frame(apply(tmp, 2, function(x) return( x / sd(x, na.rm = T))))
tmp[is.na(tmp)] <- 0
p <- pheatmap(t(tmp))
clustering <- cutree(p$tree_col, k = 2)
names(clustering) == rownames(tmp)
tmp$phenotype <- ps_mtg_filt@sam_data[samples, "phenotype"]
tmp$phenotype <- ifelse(tmp$phenotype == "A", 1, 0)

data1 <- tmp[clustering == 1, ]
data2 <- tmp[clustering == 2, ]
```


# Group by fat, veg, and fruit consumption
```{r}
ps_mtg_filt@sam_data

```
